---
title: "R Notebook"
output: html_notebook
---

## 1.1) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)
library(Seurat)
require(scales)

# Quick load
load("/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/Brandes.RData")

source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions_SeuratV3.R")
outputFolder<-"/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/"
sink(file = "/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/Brandes-AnalysisDavid-Seurat3.rmd.log", append = TRUE, split = TRUE)
```


## 1.2) Define static parameters
```{r}
#Static Parameters 

Sample.Paths <- c("/media/ATLAS_NGS_storage/Brandes_Group/103640/103640-001/outs/filtered_feature_bc_matrix",
"/media/ATLAS_NGS_storage/Brandes_Group/103640/103640-002/outs/filtered_feature_bc_matrix",
"/media/ATLAS_NGS_storage/Brandes_Group/103640/103640-003/outs/filtered_feature_bc_matrix",
"/media/ATLAS_NGS_storage/Brandes_Group/103640/103640-004/outs/filtered_feature_bc_matrix")
Samplenames <- c("Day0","Day3","Day7","Day14")
```

```{r}
#' Import Single cell sequencing experiments into Seurat3and perform normalisation and scale Data 
#' @author David John
#' @param pathways A vector of pathways to the cellrancer count output folder (contains barcodes.tsv, genes.tsv, matrix.mtx)
#' @param ids Vector of strings that are assigned to the concordant cells
#' @return Merged seurat object
Importer <- function(pathway,id, TenX=TRUE, performNormalisation=TRUE, performVariableGeneDetection=TRUE) {
  if (TenX) {
    Matrix <- Read10X(pathway)
  }  else{
    Matrix <- read.table(pathway,header = TRUE,sep = ",", dec = ".", row.names = 1)
  }
  seuratObject =CreateSeuratObject(counts = Matrix, project = id, min.cells = 5)
  seuratObject$sample <- id
  tmp<-unlist(strsplit(id,split = "-"))
  seuratObject <- subset(x = seuratObject, subset = nFeature_RNA > 500)
  if (performNormalisation==TRUE) {
    seuratObject<-NormalizeData(object = seuratObject,verbose = FALSE)
    }
  if(performVariableGeneDetection){
    seuratObject<-FindVariableFeatures(object = seuratObject, do.plot = FALSE, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  }
  #seuratObject<-ScaleData(object = seuratObject)
  cat("Imported ", length(seuratObject@meta.data$orig.ident), " cells from ", pathway, "with ID ", id, "\n")
  return(seuratObject)
}

SeuratObjectList <- list()
for (i in 1:length(Sample.Paths)) {
  SeuratObjectList[[i]]<-Importer(pathway = Sample.Paths[i],id = Samplenames[i])
}



SeuratObject.anchors <- FindIntegrationAnchors(object.list = SeuratObjectList, dims = 1:20)
SeuratObject.combined <- IntegrateData(anchorset = SeuratObject.anchors, dims = 1:20)
```


```{r}
DefaultAssay(object = SeuratObject.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
SeuratObject.combined <- ScaleData(object = SeuratObject.combined, verbose = FALSE)
SeuratObject.combined <- RunPCA(object = SeuratObject.combined, npcs = 30, verbose = FALSE)
# t-SNE and Clustering
SeuratObject.combined <- RunUMAP(object = SeuratObject.combined, reduction = "pca", dims = 1:20)
SeuratObject.combined <- FindNeighbors(object = SeuratObject.combined, reduction = "pca", dims = 1:20)
SeuratObject.combined <- FindClusters(SeuratObject.combined, resolution = 0.5)
```

```{r}
save(SeuratObject.combined,file = paste0(outputFolder,"Brandes.RData"))
```

```{r}
load(file = paste0(outputFolder,"Brandes.RData"))
```

## 3.1.5)generate TSNE Plots
####All Samples combined
```{r fig.height=10, fig.width=20}
require(cowplot)
# Visualization
p1<-DimPlot(object = SeuratObject.combined, reduction = "umap", group.by = "orig.ident",pt.size = 1)
p2<-DimPlot(object = SeuratObject.combined, reduction = "umap", label = TRUE,pt.size = 1, label.size = 9)
#p3<-DimPlot(object = SeuratObject.combined, reduction = "umap", group.by = "sample",pt.size = 2)
plot_grid(p1,p2)

DimPlot(object = SeuratObject.combined, reduction = "umap", label = TRUE, split.by = "orig.ident")

```

```{r}
FeaturePlot(object = SeuratObject.combined, features = c("Pecam1","Dcn","Vwf","Ccl5"))
```


## 3.1.6) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
library(stringr)
V<- SeuratObject.combined@meta.data
orig.ident.ordered<-str_sort(unique(SeuratObject.combined@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$seurat_clusters,levels = c(0:length(unique(SeuratObject.combined@meta.data$seurat_clusters))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= orig.ident))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()


```


##3.1.7. Find Cluster specific Markers
```{r}
SeuratObject.combined.markers <- FindAllMarkers(object = SeuratObject.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
top20<-SeuratObject.combined.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
write.csv(top20, file = paste0(outputFolder,"top20ClusterMarkers.csv"))
top20

```


## 3.2. Cell cyle analysis

```{r}
# Import human cell cycle list
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# convert list to mouse

#' Converts a human gene list into mouse
#' @author Lukas Tombor
#' @param x A vector containing a list of human genes
#' @return Converted vector
convertHumanGeneList <- function(x){
  
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  
  genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)
  
  humanx <- unique(genesV2[, 2])
  
  # Print the first 6 genes found to the screen
  print(head(humanx))
  return(humanx)
}

s.genes_mouse <- convertHumanGeneList(s.genes)
g2m.genes_mouse <- convertHumanGeneList(g2m.genes)

# Assign cell cycle scores

SeuratObject.combined <- CellCycleScoring(SeuratObject.combined, s.features = s.genes_mouse, g2m.features = g2m.genes_mouse, set.ident = T)

df <- data.frame(CCscore = SeuratObject.combined$Phase, Sample = SeuratObject.combined$sample)

df.tally <- df %>% group_by(Sample, CCscore) %>% tally() %>% mutate(percent = round(n/sum(n), digits = 4))

df.tally$Sample <- factor(df.tally$Sample, levels = c("Day0", "Day3", "Day7", "Day14"))

plot <- ggplot(df.tally, aes(x=CCscore, y= percent, fill = CCscore))+
  geom_col(color = "black")+
  scale_y_continuous(labels = scales::percent_format(), name = "% cells per Timepoint")+
  facet_grid(~Sample)+
  labs(title = "Cellcycle Score comparing Timepoints", 
       x = "Cellcycle Phase")
setwd(outputFolder)
save(SeuratObject.combined, file = "Brandes.RData")

pdf(file = "Cellcycle.pdf")
print(plot)
dev.off()
```


```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")

BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'Matrix.utils'))

library(monocle)
```



```{r}
importCDS_DJ<-function (otherCDS, import_all = FALSE) 
{
    if (class(otherCDS)[1] == "Seurat") {
        requireNamespace("Seurat")
        data <- otherCDS@assays$RNA@counts
        if (class(data) == "data.frame") {
            data <- as(as.matrix(data), "sparseMatrix")
        }
        pd <- tryCatch({
            pd <- new("AnnotatedDataFrame", data = otherCDS@meta.data)
            pd
        }, error = function(e) {
            pData <- data.frame(cell_id = colnames(data), row.names = colnames(data))
            pd <- new("AnnotatedDataFrame", data = pData)
            message("This Seurat object doesn't provide any meta data")
            pd
        })
        if (length(setdiff(colnames(data), rownames(pd))) > 0) {
            data <- data[, rownames(pd)]
        }
        fData <- data.frame(gene_short_name = row.names(data), 
            row.names = row.names(data))
        fd <- new("AnnotatedDataFrame", data = fData)
        lowerDetectionLimit <- 1
        if (all(data == floor(data))) {
            expressionFamily <- negbinomial.size()
        }
        else if (any(data < 0)) {
            expressionFamily <- uninormal()
        }
        else {
            expressionFamily <- tobit()
        }
        valid_data <- data[, row.names(pd)]
        monocle_cds <- newCellDataSet(data, phenoData = pd, featureData = fd, 
            lowerDetectionLimit = lowerDetectionLimit, expressionFamily = expressionFamily)
        if (import_all) {
            if ("Monocle" %in% names(otherCDS@misc)) {
                otherCDS@misc$Monocle@auxClusteringData$seurat <- NULL
                otherCDS@misc$Monocle@auxClusteringData$scran <- NULL
                monocle_cds <- otherCDS@misc$Monocle
                mist_list <- otherCDS
            }
            else {
                mist_list <- otherCDS
            }
        }
        else {
            mist_list <- list()
        }
        if ("var.genes" %in% slotNames(otherCDS)) {
            var.genes <- setOrderingFilter(monocle_cds, otherCDS@var.genes)
        }
        monocle_cds@auxClusteringData$seurat <- mist_list
    }
    else if (class(otherCDS)[1] == "SCESet") {
        requireNamespace("scater")
        message("Converting the exprs data in log scale back to original scale ...")
        data <- 2^otherCDS@assayData$exprs - otherCDS@logExprsOffset
        fd <- otherCDS@featureData
        pd <- otherCDS@phenoData
        experimentData = otherCDS@experimentData
        if ("is.expr" %in% slotNames(otherCDS)) 
            lowerDetectionLimit <- otherCDS@is.expr
        else lowerDetectionLimit <- 1
        if (all(data == floor(data))) {
            expressionFamily <- negbinomial.size()
        }
        else if (any(data < 0)) {
            expressionFamily <- uninormal()
        }
        else {
            expressionFamily <- tobit()
        }
        if (import_all) {
            mist_list <- otherCDS
        }
        else {
            mist_list <- list()
        }
        monocle_cds <- newCellDataSet(data, phenoData = pd, featureData = fd, 
            lowerDetectionLimit = lowerDetectionLimit, expressionFamily = expressionFamily)
        monocle_cds@auxOrderingData$scran <- mist_list
    }
    else {
        stop("the object type you want to export to is not supported yet")
    }
    return(monocle_cds)
}
```


#run Monocle trajectories for SMC's
```{r}
SeuratObject.combined.monocle<-importCDS_DJ(otherCDS = exprs(SeuratObject.combined@assays$RNA@counts, import_all = TRUE))
importCDS
```







# Monocle analysis Whole dataset
```{r}
library(monocle)

Data<-SeuratObject.combined
data <- as(as.matrix(Data@assays$RNA@data), 'sparseMatrix')

pd <- new('AnnotatedDataFrame', data = Data@meta.data)

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle cds
Data.Monocle <- newCellDataSet(data,
                               phenoData = pd,
                               featureData = fd)

Data.Monocle <- estimateSizeFactors(Data.Monocle)
Data.Monocle <- estimateDispersions(Data.Monocle)
print(head(pData(Data.Monocle)))

table(Data@meta.data$orig.ident)
Condition <- pData(Data.Monocle)$orig.ident
pData(Data.Monocle)$condition <- factor(Condition, levels = c("Day0", "Day3","Day7","Day14"))


disp_table <- dispersionTable(Data.Monocle)

Data.Monocle <- detectGenes(Data.Monocle, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(Data.Monocle), num_cells_expressed >= 10))


diff_test_res <- differentialGeneTest(Data.Monocle[expressed_genes,],
                                      fullModelFormulaStr = "~orig.ident")
print(head(diff_test_res))

ordering_genes <- row.names(subset(diff_test_res, qval < 0.000001))

print(head(ordering_genes))

Data.Monocle <- setOrderingFilter(Data.Monocle, ordering_genes)
plot_ordering_genes(Data.Monocle)

Data.Monocle <- reduceDimension(Data.Monocle, max_components = 2,
                                method = 'DDRTree')
Data.Monocle <- orderCells(Data.Monocle, reverse = T)

plot_cell_trajectory(Data.Monocle, color_by = "State", show_state_number = F, show_tree = T) +
  facet_wrap(~condition,nrow = 1) + 
  theme(axis.line = element_line(size = 6), axis.text.x = element_text(size=16, face = "italic"),
        axis.text.y = element_text(size=18),
        axis.title = element_text(size=18), legend.position = "top", strip.text = element_text(size=18))


States <- Data.Monocle@phenoData@data$State
names(States) <- rownames(Data.Monocle@phenoData@data)

Data <- AddMetaData(Data, States, "State")

Idents(Data) <- "State"
Marker.states <- FindAllMarkers(Data, test.use = "bimod", only.pos = T, logfc.threshold = 0.1)

Marker.states <- filter(Marker.states, p_val_adj < 0.05)
write.csv2(x = Marker.states, file = "/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/PseudotimeStateGenes.csv")

plot_cell_trajectory(Data.Monocle, color_by = "Pseudotime", cell_size = 1) +
  scale_color_viridis_c() + facet_wrap(~condition,nrow = 1) 


plot_cell_trajectory(Data.Monocle, markers = c("Col1a1", "Cd3d", "Lyz2"), use_color_gradient = TRUE)



```




Monocle analysis for Cluster "3",'6','13', '11', '8', '10'
```{r}
table(Idents(SeuratObject.combined))

Data<-subset(SeuratObject.combined, idents = c("3",'6','13', '11', '8', '10'))
data <- as(as.matrix(Data@assays$RNA@data), 'sparseMatrix')

pd <- new('AnnotatedDataFrame', data = Data@meta.data)

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle cds
Data.Monocle <- newCellDataSet(data,
                               phenoData = pd,
                               featureData = fd)

Data.Monocle <- estimateSizeFactors(Data.Monocle)
Data.Monocle <- estimateDispersions(Data.Monocle)
print(head(pData(Data.Monocle)))

table(Data@meta.data$orig.ident)
Condition <- pData(Data.Monocle)$orig.ident
pData(Data.Monocle)$condition <- factor(Condition, levels = c("Day0", "Day3","Day7","Day14"))


disp_table <- dispersionTable(Data.Monocle)

Data.Monocle <- detectGenes(Data.Monocle, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(Data.Monocle), num_cells_expressed >= 10))


diff_test_res <- differentialGeneTest(Data.Monocle[expressed_genes,],
                                      fullModelFormulaStr = "~orig.ident")
print(head(diff_test_res))

ordering_genes <- row.names(subset(diff_test_res, qval < 0.000001))

print(head(ordering_genes))

Data.Monocle <- setOrderingFilter(Data.Monocle, ordering_genes)
plot_ordering_genes(Data.Monocle)

Data.Monocle <- reduceDimension(Data.Monocle, max_components = 2,
                                method = 'DDRTree')
Data.Monocle <- orderCells(Data.Monocle, reverse = T)

plot_cell_trajectory(Data.Monocle, color_by = "State", show_state_number = F, show_tree = T) +
  facet_wrap(~condition,nrow = 1) + 
  theme(axis.line = element_line(size = 6), axis.text.x = element_text(size=16, face = "italic"),
        axis.text.y = element_text(size=18),
        axis.title = element_text(size=18), legend.position = "top", strip.text = element_text(size=18))


States <- Data.Monocle@phenoData@data$State
names(States) <- rownames(Data.Monocle@phenoData@data)

Data <- AddMetaData(Data, States, "State")

Idents(Data) <- "State"
Marker.states <- FindAllMarkers(Data, test.use = "bimod", only.pos = T, logfc.threshold = 0.1)

Marker.states <- filter(Marker.states, p_val_adj < 0.05)
write.csv2(x = Marker.states, file = "/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/PseudotimeStateGenes.csv")

plot_cell_trajectory(Data.Monocle, color_by = "Pseudotime", cell_size = 1) +
  scale_color_viridis_c() + facet_wrap(~condition,nrow = 1) 


plot_cell_trajectory(Data.Monocle, markers = c("Col1a1", "Cd3d", "Lyz2"), use_color_gradient = TRUE)
plot_cell_trajectory(Data.Monocle, markers = c("Acta2", "Myh11", "Cnn1", "Myl6"), use_color_gradient = TRUE)


```
#Find state specific Markers
```{r}
my_pseudotime_de <- differentialGeneTest(Data.Monocle, fullModelFormulaStr = "~sm.ns(Pseudotime)", cores = 8)
my_pseudotime_de %>% arrange(qval) %>% head()
my_pseudotime_de$id<-rownames(my_pseudotime_de)
my_pseudotime_de %>% arrange(qval) %>% head() %>% select(id) -> my_pseudotime_gene
my_pseudotime_gene <- my_pseudotime_gene$id
my_pseudotime_de %>% arrange(qval) -> my_pseudotime_de
```

```{r}
plot_genes_in_pseudotime(Data.Monocle[c("Acta2", "Myh11", "Cnn1", "Myl6"),])
plot_genes_in_pseudotime(Data.Monocle[my_pseudotime_gene,])
```


```{r}
my_branched_heatmap <- plot_genes_branched_heatmap(Data.Monocle[row.names(my_pseudotime_de[1:40,]),],
                                                   branch_point = 1,
                                                   num_clusters = 4,
                                                   cores = 8,
                                                   use_gene_short_name = TRUE,
                                                   show_rownames = TRUE,
                                                   return_heatmap = TRUE)


my_branched_heatmap <- plot_genes_branched_heatmap(Data.Monocle[row.names(my_pseudotime_de[1:40,]),],
                                                   cores = 8,
                                                   use_gene_short_name = TRUE,
                                                   show_rownames = TRUE,
                                                   return_heatmap = TRUE)

```


```{r}
options(repr.plot.width=8, repr.plot.height=12)
plot_multiple_branches_heatmap(Data.Monocle[c(positive_score_genes, negtive_score_genes),],
                               branches=c(1, 3, 4, 6, 11, 9),
                               branches_name=c("Neu", "DC", "Baso", "Mono", "MK", "Ery"),
                               show_rownames=T,
                               num_clusters=4)
```



#save workspace
```{r}
save.image("/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/Brandes.RData")
load("/media/ATLAS_NGS_storage/Brandes_Group/103640/Seurat3/Brandes.RData")

```










