---
title: "R Notebook"
author: "David John"
output: html_notebook
---
  
#1.) Unimputed Data (Both Runs Combined)
## 1.1) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)



source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/Wesley-PatientData-circulatingCells.rmd.log", append = TRUE, split = TRUE)
```


## 1.2) Define static parameters
```{r}
#Static Parameters 
#CD34 Cells excluded for now
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-010",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-011",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-014",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-016",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-017",

#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-013/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-001/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-002/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-003/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-004/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-005/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-006/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-007/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-008/outs/filtered_feature_bc_matrix/")
Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","Ctrl-CD14-Young-1","HF-7",
                             "Ctrl-Young-2","Ctrl-Young-3","Ctrl-Young-4",
                             "Ctrl-Young-5","HF-8","HF-9",
                             "HF-10","HF-11")
```

## 1.3) Impute the dataset
```{r}
imputeData(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData, cluster = 20, ncores = 20)
```


##1.4) import and combine the raw cellranger counts with CCA
```{r}
sink(file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/Wesley-PatientData-circulatingCells-rawValues.rmd.log", append = TRUE, split = TRUE)
CD31.circulatingCells <- combineSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)


CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:5, do.label=TRUE, do.fast = TRUE)

#save(CD31.circulatingCells,file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/CD31.circulatingCells.notImputed.rData")

```

## 1.5)generate TSNE Plots
####All Samples combined
```{r}
CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)


CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)


```

####Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells@cell.names
CD31.circulatingCells<-AddMetaData(object = CD31.circulatingCells,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.01, do.return=TRUE)




```

####Separated by condition colored by cluster
```{r}
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
for (ident in unique(CD31.circulatingCells@meta.data$condition)){
  #cat(a,"\n")
  cellsToUse<-grep(ident, CD31.circulatingCells@cell.names, value = TRUE)
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, label.size = 4, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE)
}
```

####Separated by Patient colored by cluster
```{r}
library(stringr)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/allCells/TSNE/png/TSNE-",ident,".png",sep="")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 1,label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
}
```

####Colored by Mutations
```{r}
condition <- c()
for (barcode in colnames(CD31.circulatingCells.CD31pos@data)){
  tmp<-unlist(strsplit(barcode,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  condition<-c(condition,cx)
}
names(condition)<-CD31.circulatingCells.CD31pos@cell.names
CD31.circulatingCells.CD31pos <- AddMetaData(object = CD31.circulatingCells.CD31pos, metadata = condition, col.name = "condition")
CD31.circulatingCells.CD31pos <- SetAllIdent(object = CD31.circulatingCells.CD31pos, id = "condition")

TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2.1)
TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2.1, group.by="orig.ident", colors.use = c("green","green","green","grey","blue","blue","blue","blue","blue","red","orange","orange","pink","orange","orange","orange","orange","pink","green","green"))




```


##1.6) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
V<- CD31.circulatingCells@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:15))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

#svg(filename = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/allCells/Barplot-CellsperClusterPerSample.svg",width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
#dev.off()


```


##1.7) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

#CD31 positive
CD31.circulatingCells.CD31pos <- SetAllIdent(object = CD31.circulatingCells.CD31pos, id = "res.0.6")
CD31.circulatingCells.CD31pos.markers <- FindAllMarkers(object = CD31.circulatingCells.CD31pos, only.pos = TRUE)
dink<-CD31.circulatingCells.CD31pos.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/ClusterSpecificMarkers.csv")
```



##1.6) Only CD31 positive cells
####load saved dataset
```{r}
load(file = "/media/Storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/CD31.circulatingCells.notImputed.rData")
```


####Subset CD31 positiv cells
```{r}
CD31.circulatingCells.CD31pos<- SubsetData(CD31.circulatingCells,subset.name = "PECAM1", accept.low = 0.00001,)
length(CD31.circulatingCells.CD31pos@meta.data$orig.ident)
length(CD31.circulatingCells@meta.data$orig.ident)
diff<-length(CD31.circulatingCells@meta.data$orig.ident)-length(CD31.circulatingCells.CD31pos@meta.data$orig.ident)
cat(diff," Cells were PECAM1 negativ")

```

####rerun  clustering
```{r}
CD31.circulatingCells.CD31pos <- FindVariableGenes(object = CD31.circulatingCells.CD31pos)
CD31.circulatingCells.CD31pos <- RunPCA(object = CD31.circulatingCells.CD31pos, features = VariableFeatures(object = CD31.circulatingCells.CD31pos), verbose = TRUE)
CD31.circulatingCells.CD31pos<-ProjectPCA(object = CD31.circulatingCells.CD31pos, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.CD31pos, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.CD31pos <- FindClusters(object = CD31.circulatingCells.CD31pos, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.CD31pos)
CD31.circulatingCells.CD31pos <- RunTSNE(object = CD31.circulatingCells.CD31pos, dims.use = 1:5, do.label=TRUE, do.fast = TRUE)

save(CD31.circulatingCells.CD31pos,file ="/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/CD31.circulatingCells.notImputed.CD31pos.rData")
```


```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = c("PECAM1", "CD3D","CDH5","CD14","CD79A"))
VlnPlot(object = CD31.circulatingCells.CD31pos, features.plot = c("PECAM1", "CD3D","CDH5","CD14","CD79A"))
```


#### (CD31+) Generate separate TSNEs for each sample
```{r}
library(stringr)
CD31.circulatingCells.CD31pos<-SetAllIdent(CD31.circulatingCells.CD31pos,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells.CD31pos@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells.CD31pos<-SetAllIdent(CD31.circulatingCells.CD31pos,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells.CD31pos@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/TSNE/png/TSNE-",ident,".png",sep="")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 1, label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
}
```

#### (CD31+) Generate barplot of cell per cluster
```{r fig.height=30, fig.width=20}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
V<- CD31.circulatingCells.CD31pos@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells.CD31pos@meta.data$orig.ident),numeric = TRUE)


V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:15))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

#svg(filename = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/Barplot-CellsperClusterPerSample.svg",width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
#dev.off()
```






##1.7) perform pre analysis
this analysis is only for stefanie to confirm the previous results. 
```{r}
Path.PatientData.CD31.circulatingCells <- c(
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-013/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/"
)

Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","Ctrl-CD14-Young-1","HF-7"
)

#import and combine the new samples
CD31.circulatingCells <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)

CD31.circulatingCells<-NormalizeData(object = CD31.circulatingCells)


#regress out batch effects 
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CD31.circulatingCells@data), value = TRUE)
percent.mito <- Matrix::colSums(CD31.circulatingCells@raw.data[mito.genes, ])/Matrix::colSums(CD31.circulatingCells@raw.data)
CD31.circulatingCells <- AddMetaData(object = CD31.circulatingCells, metadata = percent.mito, col.name = "percent.mito");rm(mito.genes, percent.mito)
CD31.circulatingCells.filtered <- FilterCells(object = CD31.circulatingCells, subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(Inf, 0.1))


GenePlot(object = CD31.circulatingCells, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = CD31.circulatingCells.filtered, gene1 = "nUMI", gene2 = "percent.mito")


CD31.circulatingCells.filtered <- FindVariableGenes(object = CD31.circulatingCells.filtered)
CD31.circulatingCells.filtered <- RunPCA(object = CD31.circulatingCells.filtered, features = VariableFeatures(object = CD31.circulatingCells.filtered), verbose = TRUE)
CD31.circulatingCells.filtered<-ProjectPCA(object = CD31.circulatingCells.filtered, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.filtered, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.filtered <- FindClusters(object = CD31.circulatingCells.filtered, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.filtered)
CD31.circulatingCells.filtered <- RunTSNE(object = CD31.circulatingCells.filtered, dims.use = 1:5, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 2.1, do.return=TRUE)

TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 2.1, group_by = "orig.ident", do.label = T)

CD31.circulatingCells.filtered<-SetAllIdent(CD31.circulatingCells.filtered, id = "orig.ident")


plotList<- c()
for (ident in sort(unique(CD31.circulatingCells.filtered@meta.data$orig.ident))[1:4]) {
  cat(i,"\n")
  
  #plotList<-c(plotList,ident)
  #svg(filename = paste("/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TNSE",ident,".svg",sep = "-"))
  #cellToUse<-c(cellToUse,grep(c("HF-1","HF-2","HF-3"), CD31.circulatingCells.filtered@cell.names, value = TRUE))
  #TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 0.1, group_by = "orig.ident", do.label = T, label.size = 5, 
  #cells.use = grep(c("HF-1","HF-2","HF-3"), CD31.circulatingCells.filtered@cell.names, value = TRUE))
  
  #dev.off()
}

ggarrange(plotlist = plotList[1])

```


#2.) Imputed Data (Both Runs Combined)
this analysis is only for stefanie to confirm the previous results. 
```{r}
CD31.circulatingCells <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)
CD31.circulatingCells<-NormalizeData(object = CD31.circulatingCells)



#regress out batch effects 
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CD31.circulatingCells@data), value = TRUE)
percent.mito <- Matrix::colSums(CD31.circulatingCells@raw.data[mito.genes, ])/Matrix::colSums(CD31.circulatingCells@raw.data)
CD31.circulatingCells <- AddMetaData(object = CD31.circulatingCells, metadata = percent.mito, col.name = "percent.mito");rm(mito.genes, percent.mito)
CD31.circulatingCells.filtered <- FilterCells(object = CD31.circulatingCells, subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(Inf, 0.1))


GenePlot(object = CD31.circulatingCells, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = CD31.circulatingCells.filtered, gene1 = "nUMI", gene2 = "percent.mito")


CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:5, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)

TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 0.1, group_by = "orig.ident", do.label = T)

CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells, id = "orig.ident")


cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(1,2,3)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-CTRL-AGED.svg")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()

cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(5,6,7,8,9)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-CTRL-YOUNG.svg")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()

cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(10,11,12,13,14,15,16,17,18,19,20)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-HF.svg")

TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()


ggarrange(plotlist = plotList[1])


```


#3.)Unimputed Data (Separated by Run)
## 3.1) First Run
###3.1.1) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)

logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.rmd.log"
if (file.exists(logFile)) {
  #file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)
```


## 1.2) Define static parameters
```{r}
#Static Parameters 
#CD34 Cells excluded for now
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-010",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-011",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-014",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-016",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-017",

#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-001/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-002/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-003/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-004/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-005/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-006/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-007/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-008/outs/filtered_feature_bc_matrix/")
Samplenames.PatientData <- c("Ctrl-Young-2","Ctrl-Young-3","Ctrl-Young-4",
                             "Ctrl-Young-5","HF-8","HF-9",
                             "HF-10","HF-11")
```




## 3.2) Second Run
###3.1.1) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)

logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019-Seurat3//Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells-Seurat3.unimputed.SeparatedByRun.SecondRun.rmd.log"
if (file.exists(logFile)) {
  file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)

```


## 3.1.2) Define static parameters
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019-Seurat3//Unimputed/SeparatedByRun/SecondRun/" 



#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
  
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
  "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/"
)
Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","HF-7"
)
```

## 3.1.3) Impute the dataset
```{r}
#imputeData(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData, cluster = 20, ncores = 20)
```

## 3.1.4) import and combine the raw cellranger counts with CCA
```{r}
#' Import Single cell sequencing experiments into Seurat3and perform normalisation and scale Data 
#' @author David John
#' @param pathways A vector of pathways to the cellrancer count output folder (contains barcodes.tsv, genes.tsv, matrix.mtx)
#' @param ids Vector of strings that are assigned to the concordant cells
#' @return Merged seurat object
Importer <- function(pathway,id, TenX=TRUE, performNormalisation=TRUE, performVariableGeneDetection=TRUE) {
  if (TenX) {
    Matrix <- Read10X(pathway)
  }  else{
    Matrix <- read.table(pathway,header = TRUE,sep = ",", dec = ".", row.names = 1)
  }
  seuratObject =CreateSeuratObject(counts = Matrix, project = id, min.cells = 5)
  seuratObject$sample <- id
  tmp<-unlist(strsplit(idrom,split = "-"))
  seuratObject$condition <- paste0(tmp[1:length(tmp)-1],collapse = "-")
  seuratObject <- subset(x = seuratObject, subset = nFeature_RNA > 500)
  if (performNormalisation==TRUE) {
    seuratObject<-NormalizeData(object = seuratObject,verbose = FALSE)
    }
  if(performVariableGeneDetection){
    seuratObject<-FindVariableFeatures(object = seuratObject, do.plot = FALSE, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  }
  #seuratObject<-ScaleData(object = seuratObject)
  cat("Imported ", length(seuratObject@meta.data$orig.ident), " cells from ", pathway, "with ID ", id, "\n")
  return(seuratObject)
}

SeuratObjectList <- list()
for (i in 1:length(Path.PatientData.CD31.circulatingCells)) {
  SeuratObjectList[[i]]<-Importer(pathway = Path.PatientData.CD31.circulatingCells[i],id = Samplenames.PatientData[i])
}



CD31.anchors <- FindIntegrationAnchors(object.list = SeuratObjectList, dims = 1:20)
CD31.combined <- IntegrateData(anchorset = CD31.anchors, dims = 1:20)
```


```{r}
DefaultAssay(object = CD31.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
CD31.combined <- ScaleData(object = CD31.combined, verbose = FALSE)
CD31.combined <- RunPCA(object = CD31.combined, npcs = 30, verbose = FALSE)
# t-SNE and Clustering
CD31.combined <- RunUMAP(object = CD31.combined, reduction = "pca", dims = 1:20)
CD31.combined <- FindNeighbors(object = CD31.combined, reduction = "pca", dims = 1:20)
CD31.combined <- FindClusters(CD31.combined, resolution = 0.5)
```


```{r}
save(CD31.circulatingCells,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.RData"))
```

## 3.1.5)generate TSNE Plots
####All Samples combined
```{r fig.height=20, fig.width=20}
require(cowplot)
# Visualization
DimPlot(object = CD31.combined, reduction = "umap", group.by = "condition",pt.size = 2)
DimPlot(object = CD31.combined, reduction = "umap", label = TRUE,pt.size = 2)
DimPlot(object = CD31.combined, reduction = "umap", group.by = "sample",pt.size = 2)
plot_grid(p1, p2,p3)

DimPlot(object = CD31.combined, reduction = "umap", label = TRUE, split.by = "condition")
p1

```





## 3.1.6) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
library(stringr)
V<- CD31.combined@meta.data
orig.ident.ordered<-str_sort(unique(CD31.combined@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$seurat_clusters,levels = c(0:length(unique(CD31.combined@meta.data$seurat_clusters))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()


```

####Save File
```{r}
save(CD31.combined,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.RData"))
```



## 3.1.7) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

## 3.1.8) DEGs HF vs CTRL-Aged
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "HF", ident.2 = "Ctrl-Aged")
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```


## 3.1.9) DEG Old vs Young 
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```