---
title: "R Notebook"
author: "David John"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

#1.) Unimputed Data (Both Runs Combined)
## 1.1.) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)



source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/Wesley-PatientData-circulatingCells.rmd.log", append = TRUE, split = TRUE)
```


## 1.2.) Define static parameters
```{r}
#Static Parameters 
#CD34 Cells excluded for now
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-010",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-011",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-014",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-016",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-017",

#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-013/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-008/outs/filtered_feature_bc_matrix/")
Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","Ctrl-CD14-Young-1","HF-7",
                             "Ctrl-Young-2","Ctrl-Young-3","Ctrl-Young-4",
                             "Ctrl-Young-5","HF-8","HF-9",
                             "HF-10","HF-11")
```

## 1.3.) Impute the dataset
```{r}
imputeData(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData, cluster = 20, ncores = 20)
```


##1.4.) import and combine the raw cellranger counts with CCA
```{r}
sink(file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/Wesley-PatientData-circulatingCells-rawValues.rmd.log", append = TRUE, split = TRUE)
CD31.circulatingCells <- combineSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)


CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:5, do.label=TRUE, do.fast = TRUE)

#save(CD31.circulatingCells,file = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/CD31.circulatingCells.notImputed.rData")

```

## 1.5.)generate TSNE Plots
All Samples combined
```{r}
CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)


CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)


```

Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells@cell.names
CD31.circulatingCells<-AddMetaData(object = CD31.circulatingCells,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.01, do.return=TRUE)




```

Separated by condition colored by cluster
```{r}
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
for (ident in unique(CD31.circulatingCells@meta.data$condition)){
  #cat(a,"\n")
  cellsToUse<-grep(ident, CD31.circulatingCells@cell.names, value = TRUE)
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, label.size = 4, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE)
  }
```

Separated by Patient colored by cluster
```{r}
library(stringr)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/allCells/TSNE/png/TSNE-",ident,".png",sep="")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 1,label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
  }
```

Colored by Mutations
```{r}
condition <- c()
for (barcode in colnames(CD31.circulatingCells.CD31pos@data)){
  tmp<-unlist(strsplit(barcode,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  condition<-c(condition,cx)
}
names(condition)<-CD31.circulatingCells.CD31pos@cell.names
CD31.circulatingCells.CD31pos <- AddMetaData(object = CD31.circulatingCells.CD31pos, metadata = condition, col.name = "condition")
CD31.circulatingCells.CD31pos <- SetAllIdent(object = CD31.circulatingCells.CD31pos, id = "condition")

TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2.1)
TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2.1, group.by="orig.ident", colors.use = c("green","green","green","grey","blue","blue","blue","blue","blue","red","orange","orange","pink","orange","orange","orange","orange","pink","green","green"))

 


```


##1.6.) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
V<- CD31.circulatingCells@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:15))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

#svg(filename = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/allCells/Barplot-CellsperClusterPerSample.svg",width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
#dev.off()


```


##1.7.) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

#CD31 positive
CD31.circulatingCells.CD31pos <- SetAllIdent(object = CD31.circulatingCells.CD31pos, id = "res.0.6")
CD31.circulatingCells.CD31pos.markers <- FindAllMarkers(object = CD31.circulatingCells.CD31pos, only.pos = TRUE)
dink<-CD31.circulatingCells.CD31pos.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/ClusterSpecificMarkers.csv")
```



##1.6.) Only CD31 positive cells


Subset CD31 positiv cells
```{r}
CD31.circulatingCells.CD31pos<- SubsetData(CD31.circulatingCells,subset.name = "PECAM1", accept.low = 0.00001,)
length(CD31.circulatingCells.CD31pos@meta.data$orig.ident)
length(CD31.circulatingCells@meta.data$orig.ident)
diff<-length(CD31.circulatingCells@meta.data$orig.ident)-length(CD31.circulatingCells.CD31pos@meta.data$orig.ident)
cat(diff," Cells were PECAM1 negativ")

```

rerun  clustering
```{r}
CD31.circulatingCells.CD31pos <- FindVariableGenes(object = CD31.circulatingCells.CD31pos)
CD31.circulatingCells.CD31pos <- RunPCA(object = CD31.circulatingCells.CD31pos, features = VariableFeatures(object = CD31.circulatingCells.CD31pos), verbose = TRUE)
CD31.circulatingCells.CD31pos<-ProjectPCA(object = CD31.circulatingCells.CD31pos, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.CD31pos, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.CD31pos <- FindClusters(object = CD31.circulatingCells.CD31pos, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.CD31pos)
CD31.circulatingCells.CD31pos <- RunTSNE(object = CD31.circulatingCells.CD31pos, dims.use = 1:5, do.label=TRUE, do.fast = TRUE)
```

#save
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/BothRunsCombined/allCells/" 
save(CD31.circulatingCells.CD31pos,file =paste(outputFolder,"CD31.circulatingCells.notImputed.rData"))
```
load
```{r}
load(file = paste(outputFolder,"CD31.circulatingCells.notImputed.rData"))
```


```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = c("PECAM1", "CD3D","CDH5","CD14","CD79A"))
VlnPlot(object = CD31.circulatingCells.CD31pos, features.plot = c("PECAM1", "CD3D","CDH5","CD14","CD79A"))
```


(CD31+) Generate separate TSNEs for each sample
```{r}
library(stringr)
CD31.circulatingCells.CD31pos<-SetAllIdent(CD31.circulatingCells.CD31pos,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells.CD31pos@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells.CD31pos<-SetAllIdent(CD31.circulatingCells.CD31pos,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells.CD31pos@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/TSNE/png/TSNE-",ident,".png",sep="")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells.CD31pos, label=T, pt.size = 1, label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
  }
```

(CD31+) Generate barplot of cell per cluster
```{r fig.height=30, fig.width=20}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
V<- CD31.circulatingCells.CD31pos@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells.CD31pos@meta.data$orig.ident),numeric = TRUE)


V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:15))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

#svg(filename = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/CD31+/Barplot-CellsperClusterPerSample.svg",width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
#dev.off()
```






##1.7.) perform pre analysis
this analysis is only for stefanie to confirm the previous results. 
```{r}
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-013/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/"
)

Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","Ctrl-CD14-Young-1","HF-7"
                             )

#import and combine the new samples
CD31.circulatingCells <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)

CD31.circulatingCells<-NormalizeData(object = CD31.circulatingCells)


#regress out batch effects 
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CD31.circulatingCells@data), value = TRUE)
percent.mito <- Matrix::colSums(CD31.circulatingCells@raw.data[mito.genes, ])/Matrix::colSums(CD31.circulatingCells@raw.data)
CD31.circulatingCells <- AddMetaData(object = CD31.circulatingCells, metadata = percent.mito, col.name = "percent.mito");rm(mito.genes, percent.mito)
CD31.circulatingCells.filtered <- FilterCells(object = CD31.circulatingCells, subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(Inf, 0.1))


GenePlot(object = CD31.circulatingCells, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = CD31.circulatingCells.filtered, gene1 = "nUMI", gene2 = "percent.mito")


CD31.circulatingCells.filtered <- FindVariableGenes(object = CD31.circulatingCells.filtered)
CD31.circulatingCells.filtered <- RunPCA(object = CD31.circulatingCells.filtered, features = VariableFeatures(object = CD31.circulatingCells.filtered), verbose = TRUE)
CD31.circulatingCells.filtered<-ProjectPCA(object = CD31.circulatingCells.filtered, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.filtered, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.filtered <- FindClusters(object = CD31.circulatingCells.filtered, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.filtered)
CD31.circulatingCells.filtered <- RunTSNE(object = CD31.circulatingCells.filtered, dims.use = 1:5, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 2.1, do.return=TRUE)

TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 2.1, group_by = "orig.ident", do.label = T)

CD31.circulatingCells.filtered<-SetAllIdent(CD31.circulatingCells.filtered, id = "orig.ident")


plotList<- c()
for (ident in sort(unique(CD31.circulatingCells.filtered@meta.data$orig.ident))[1:4]) {
  cat(i,"\n")
<<<<<<< HEAD
=======
  
>>>>>>> fd0db9047a7e7327a12ca64407617b405c357330
  #plotList<-c(plotList,ident)
  #svg(filename = paste("/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TNSE",ident,".svg",sep = "-"))
  #cellToUse<-c(cellToUse,grep(c("HF-1","HF-2","HF-3"), CD31.circulatingCells.filtered@cell.names, value = TRUE))
  #TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 0.1, group_by = "orig.ident", do.label = T, label.size = 5, 
                          #cells.use = grep(c("HF-1","HF-2","HF-3"), CD31.circulatingCells.filtered@cell.names, value = TRUE))
<<<<<<< HEAD
=======

>>>>>>> fd0db9047a7e7327a12ca64407617b405c357330
  #dev.off()
  }

ggarrange(plotlist = plotList[1])

```


#2.) Imputed Data (Both Runs Combined)
this analysis is only for stefanie to confirm the previous results. 
```{r}
CD31.circulatingCells <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)
CD31.circulatingCells<-NormalizeData(object = CD31.circulatingCells)



#regress out batch effects 
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CD31.circulatingCells@data), value = TRUE)
percent.mito <- Matrix::colSums(CD31.circulatingCells@raw.data[mito.genes, ])/Matrix::colSums(CD31.circulatingCells@raw.data)
CD31.circulatingCells <- AddMetaData(object = CD31.circulatingCells, metadata = percent.mito, col.name = "percent.mito");rm(mito.genes, percent.mito)
CD31.circulatingCells.filtered <- FilterCells(object = CD31.circulatingCells, subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(Inf, 0.1))


GenePlot(object = CD31.circulatingCells, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = CD31.circulatingCells.filtered, gene1 = "nUMI", gene2 = "percent.mito")


CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:5, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:5, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)

TSNEPlot(object = CD31.circulatingCells.filtered, label=T, pt.size = 0.1, group_by = "orig.ident", do.label = T)

CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells, id = "orig.ident")


cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(1,2,3)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-CTRL-AGED.svg")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()

cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(5,6,7,8,9)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-CTRL-YOUNG.svg")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()

cellsToUse<-c()
for (ident in sort(unique(CD31.circulatingCells@meta.data$orig.ident))[c(10,11,12,13,14,15,16,17,18,19,20)]) {cellsToUse<-c(cellsToUse,grep(ident, CD31.circulatingCells@cell.names, value = TRUE))}
svg(filename = "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/TSNE-HF.svg")

TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, group_by = "orig.ident", label.size = 5, cells.use = cellsToUse);dev.off()


ggarrange(plotlist = plotList[1])


```


#3.)Unimputed Data (Separated by Run)
## 3.1.) First Run
###3.1.1.) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)
library(Seurat)
logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.rmd.log"
if (file.exists(logFile)) {
  #file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)
```


<<<<<<< HEAD
## 3.1.2.) Define static parameters
=======
```{r}
#Static Parameters 
#CD34 Cells excluded for now
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-010",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-011",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-014",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-016",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-017",

#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-008/outs/filtered_feature_bc_matrix/")
Samplenames.PatientData <- c("Ctrl-Young-2","Ctrl-Young-3","Ctrl-Young-4",
                             "Ctrl-Young-5","HF-8","HF-9",
                             "HF-10","HF-11")
```

## 3.1.3.) import and combine the raw cellranger counts with CCA
```{r}
CD31.circulatingCells <- combineSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)

CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:11, resolution = 0.6, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:11, do.label=TRUE, do.fast = TRUE)
```

## 3.1.4.)generate TSNE Plots
All Samples combined
```{r}
CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE, do.label = TRUE, label.size = 9)


CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)
```

Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells@cell.names
CD31.circulatingCells<-AddMetaData(object = CD31.circulatingCells,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.01, do.return=TRUE)
```

Separated by condition colored by cluster
```{r}
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
for (ident in unique(CD31.circulatingCells@meta.data$condition)){
  #cat(a,"\n")
  cellsToUse<-grep(ident, CD31.circulatingCells@cell.names, value = TRUE)
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, label.size = 4, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE)
  }
```

Separated by Patient colored by cluster
```{r}
library(stringr)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells@cell.names, value = TRUE)
  #head(cellsToUse)
  #filename <- paste0(outputFolder,"/TSNE/png/TSNE-",ident,".png")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 1,label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
  }
```

Colored by Mutations
```{r}
TSNEPlot(object = CD31.circulatingCells.s, label=T, pt.size = 2.1, group.by="orig.ident", colors.use = c("green","green","green","grey","blue","blue","blue","blue","blue","red","orange","orange","pink","orange","orange","orange","orange","pink","green","green"))
```


## 3.1.5.) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
library(stringr)
V<- CD31.circulatingCells@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:length(unique(CD31.circulatingCells@meta.data$res.0.6))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()


```

##save
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/" 
save(CD31.circulatingCells,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.RData"))
```

```{r}

load(file=paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.RData"))
```

## 3.1.6.) DEG's
### 3.1.6.1.) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

generate vlnPlots
```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = din)

```


### 3.1.6.2.) DEGs HF vs CTRL-Aged
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.CtrlAgedvsHF.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "Ctrl-Aged", ident.2 = "HF")
# --> No DEGs found

CD31.circulatingCells.CtrlAgedvsHF.markers$geneName<-rownames(CD31.circulatingCells.CtrlAgedvsHF.markers)
CD31.circulatingCells.CtrlAgedvsHF.markers<-CD31.circulatingCells.CtrlAgedvsHF.markers[order(CD31.circulatingCells.CtrlAgedvsHF.markers$avg_logFC),]
topup<-head(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)
topdown<-tail(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)


write.csv(CD31.circulatingCells.CtrlAgedvsHF.markers,file = paste0(outputFolder,"ClusterSpecificMarkers.CtrlAged-vs-HF.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```


```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells_downregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "Ctrl-Aged", ident.2 = "HF", only.pos = TRUE,print.bar = T)
CD31.circulatingCells_upregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "HF", ident.2 = "Ctrl-Aged", only.pos = TRUE,print.bar = T)
head(CD31.circulatingCells_downregulated,n=9);head(CD31.circulatingCells_upregulated,n=9)
topDown<-head(CD31.circulatingCells_downregulated,n=6)
topUp<-head(CD31.circulatingCells_upregulated,n=6)

```

generate vlnPlots topUp colored by orig.ident
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")

VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topUp))
VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topDown))
```



# _
### 3.1.7.) CircRes Paper

##save
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/" 
save(CD31.circulatingCells,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.RData"))
```

```{r}

load(file=paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.RData"))
CD31.circulatingCells.Cohort1<-CD31.circulatingCells
```

#####request from Stefanie from 05.06.19 19:08 Expression of TREM2 on Cohort1 
```{r}
CD31.circulatingCells.Cohort1 <- SetAllIdent(CD31.circulatingCells.Cohort1, id = "condition")
FeaturePlot(CD31.circulatingCells.Cohort1,features=c("TREM2"))
VlnPlot(CD31.circulatingCells,features=c("TREM2"))
```



select CD31+ cells
```{r}
CircRes_Cohort1_CD31Pos <- SubsetData(CD31.circulatingCells.Cohort1, subset.name = "PECAM1", accept.low = 0.001)
CircRes_Cohort1_CD31Pos <- FindVariableGenes(object = CircRes_Cohort1_CD31Pos, do.recalc = T)
CircRes_Cohort1_CD31Pos <- RunPCA(object = CircRes_Cohort1_CD31Pos, features = VariableFeatures(object = CircRes_Cohort1_CD31Pos), verbose = TRUE)
CircRes_Cohort1_CD31Pos<-ProjectPCA(object = CircRes_Cohort1_CD31Pos, do.print = FALSE)
PCHeatmap(object = CircRes_Cohort1_CD31Pos, pc.use = 1:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CircRes_Cohort1_CD31Pos <- FindClusters(object = CircRes_Cohort1_CD31Pos, reduction.type = "pca", dims.use = 1:7, resolution = 0.6, save.SNN = TRUE, force.recalc = T)
PrintFindClustersParams(object = CircRes_Cohort1_CD31Pos)
CircRes_Cohort1_CD31Pos <- RunTSNE(object = CircRes_Cohort1_CD31Pos, dims.use = 1:4, do.label=TRUE)
```

```{r}

newSamplename<-c()
for (sampleName in CircRes_Cohort1_CD31Pos@meta.data$orig.ident) {
  if (sampleName=="HF-1") {newSamplename<-c(newSamplename,"HF2.1")}
  if (sampleName=="HF-2") {newSamplename<-c(newSamplename,"HF2.2")}
  if (sampleName=="HF-3") {newSamplename<-c(newSamplename,"HF2.3")}
  if (sampleName=="HF-4") {newSamplename<-c(newSamplename,"HF2.4")}
  if (sampleName=="HF-5") {newSamplename<-c(newSamplename,"HF2.5")}
  if (sampleName=="HF-6") {newSamplename<-c(newSamplename,"HF2.6")}
  if (sampleName=="HF-7") {newSamplename<-c(newSamplename,"HF2.7")}
  if (sampleName=="HF-8") {newSamplename<-c(newSamplename,"HF1.1")}
  if (sampleName=="HF-9") {newSamplename<-c(newSamplename,"HF1.2")}
  if (sampleName=="HF-10") {newSamplename<-c(newSamplename,"HF1.3")}
  if (sampleName=="HF-11") {newSamplename<-c(newSamplename,"HF1.4")}
  if (sampleName=="Ctrl-Aged-1") {newSamplename<-c(newSamplename,"HC2.1")}
  if (sampleName=="Ctrl-Aged-2") {newSamplename<-c(newSamplename,"HC2.2")}
  if (sampleName=="Ctrl-Aged-3") {newSamplename<-c(newSamplename,"HC2.3")}
  if (sampleName=="Ctrl-Young-1") {newSamplename<-c(newSamplename,"HCY2.1")}
  if (sampleName=="Ctrl-Young-2") {newSamplename<-c(newSamplename,"HCY1.1")}
  if (sampleName=="Ctrl-Young-3") {newSamplename<-c(newSamplename,"HCY1.2")}
  if (sampleName=="Ctrl-Young-4") {newSamplename<-c(newSamplename,"HCY1.3")}
  if (sampleName=="Ctrl-Young-5") {newSamplename<-c(newSamplename,"HCY1.4")}
  if (sampleName=="Ctrl-CD14-Young-1") {newSamplename<-c(newSamplename,"HCY2.1_CD14")}
  else {cat("unknown Samplename: ", sampleName, "\n")}
}
table(newSamplename)
names(newSamplename)<-CircRes_Cohort1_CD31Pos@cell.names
CircRes_Cohort1_CD31Pos<-AddMetaData(object = CircRes_Cohort1_CD31Pos,metadata = data.frame(newSamplename),col.name = "newSamplenames")
table(CircRes_Cohort1_CD31Pos@meta.data$newSamplename)

rm(newSamplename)
```

##save
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/" 
save(CircRes_Cohort1_CD31Pos,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.CD31+.CircRes.RData"))
```

```{r}
load(file=paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.FirstRun.CD31+.CircRes.RData"))
```

plot Monocyte markers 
```{r}
#CircRes_Cohort2<-SetAllIdent(object = CircRes_Cohort2, id = "res.0.6")
VlnPlot(CircRes_Cohort1_CD31Pos, features.plot = "PECAM1")
FeaturePlot(object = CircRes_Cohort1_CD31Pos,features.plot = c("FUT4","ELANE","CEACAM1"))
TSNEPlot(CircRes_Cohort1_CD31Pos, do.label = T)

CircRes_Cohort1_CD31Pos <- SetAllIdent(object = CircRes_Cohort1_CD31Pos, id = "newSamplename")
VlnPlot(CircRes_Cohort1_CD31Pos, features.plot = c("FUT4","ELANE"), x.lab.rot = 90)
FeaturePlot(object = CircRes_Cohort1_CD31Pos,features.plot = c("TREM2"))
```


#####request from Stefanie from 05.06.19 19:08 Expression of TREM2 on Cohort1 
```{r}
CircRes_Cohort1_CD31Pos <- SetAllIdent(CircRes_Cohort1_CD31Pos, id = "condition")
FeaturePlot(CircRes_Cohort1_CD31Pos,features=c("TREM2"))
VlnPlot(CircRes_Cohort1_CD31Pos,features=c("TREM2"))


#Get percentage of Neutrophil Marker positive cells
```{r}
df.NeutrophilPos.Cohort1 <- data.frame()
CircRes_Cohort1_CD31Pos<-SetAllIdent(CircRes_Cohort1_CD31Pos,id = "orig.ident")
for (x in unique(CircRes_Cohort1_CD31Pos@meta.data$orig.ident)) {
  cat(x,"\n")
  tmp<-SubsetData(CircRes_Cohort1_CD31Pos,ident.use = x)
  df.NeutrophilPos[x,"TotalCells"]<-length(tmp@ident)
 
  n<-length(which(tmp@data["FUT4",]!=0))
  df.NeutrophilPos[x,"Number_Fut4_positive"]<-n
  df.NeutrophilPos[x,"Percentage_Fut4_positive"] <- (n/length(tmp@ident))*100
  
  n<-length(which(tmp@data["ELANE",]!=0))
  df.NeutrophilPos[x,"Number_ELANE_positive"]<-n
  df.NeutrophilPos[x,"Percentage_ELANE_positive"] <- (n/length(tmp@ident))*100

  n<-length(which(tmp@data["CEACAM1",]!=0))
  df.NeutrophilPos[x,"Number_CEACAM_positive"]<-n
  df.NeutrophilPos[x,"Percentage_CEACAM_positive"] <- (n/length(tmp@ident))*100
}
write.csv(x = df.NeutrophilPos, file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/PercentageNeutrophilMarkers.CircRes.csv")
```

### 3.1.7)Run Scenic for Cohort 1
```{r}
dbFiles <- c("https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-500bp-upstream-7species.mc9nr.feather",
"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather")
setwd("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/FirstRun/SCENIC")

for(featherURL in dbFiles)
{
  download.file(featherURL, destfile=basename(featherURL)) # saved in current dir
  descrURL <- gsub(".feather$", ".descr", featherURL)
  if(file.exists(descrURL)) download.file(descrURL, destfile=basename(descrURL))
}

pfile <- Convert(from = CD31.circulatingCells, to = "loom", filename = "CD31.circulatingCells.loom", 
    display.progress = TRUE)


```

#TODO
plot characktristic Monocyte markers (e.g. CD14, FCGRCA, CD64, CD68, CD163, CXC3R1, CCR5, CCR2, CD192,IL1, IL10, TLR2, TLR3,)
Barplot with error marks + Raw Data in Table
```{r}
BothCohortsMonocytes<-readRDS("/media/ATLAS_NGS_storage/Wesley/CircRes2019/CircResBothCohortsMonocytes.final.Rds")
BothCohortsMonocytes<-SetAllIdent(BothCohortsMonocytes,id = "Monocyte.Annotations")
#TSNEPlot(BothCohortsMonocytes)
table(BothCohortsMonocytes@ident)
Monocyte.Markers <- c("CD14", "IL1B", "CD36", "TLR2", "FCGR1A", "TNFRSF1A","CD163", "CCR2","CD68", "FCGR3A", "IL1A", "IL10", "TLR2", "TLR3")
Neutrophil.Markers <- c("FUT4", "ELANE", "CEACAM1")

sem <- function(x) sd(x)/sqrt(length(x)) #define function to calc SEM


#"Classical.Monocyte"     "Intermediate.Monocyte"  "Non.Classical.Monocyte"

BothCohortsMonocytes<-SetAllIdent(BothCohortsMonocytes,id = "Monocyte.Annotations")
Classical.Monocyte<-SubsetData(BothCohortsMonocytes, ident.use = "Classical.Monocyte")
Classical.Monocyte<-SetAllIdent(Classical.Monocyte,id = "Conditions")
Classical.Monocyte.HC_HCY <- SubsetData(Classical.Monocyte, ident.use = c("HC","HCY"))
Classical.Monocyte.HF <- SubsetData(Classical.Monocyte, ident.use = c("HF"))


Intermediate.Monocyte<-SubsetData(BothCohortsMonocytes, ident.use = "Intermediate.Monocyte")
Intermediate.Monocyte<-SetAllIdent(Intermediate.Monocyte,id = "Conditions")
Intermediate.Monocyte.HC_HCY <- SubsetData(Intermediate.Monocyte, ident.use = c("HC","HCY"))
Intermediate.Monocyte.HF <- SubsetData(Intermediate.Monocyte, ident.use = c("HF"))

Non.Classical.Monocyte<-SubsetData(BothCohortsMonocytes, ident.use = "Non.Classical.Monocyte")
Non.Classical.Monocyte<-SetAllIdent(Non.Classical.Monocyte,id = "Conditions")
Non.Classical.Monocyte.HC_HCY <- SubsetData(Non.Classical.Monocyte, ident.use = c("HC","HCY"))
Non.Classical.Monocyte.HF <- SubsetData(Non.Classical.Monocyte, ident.use = c("HF"))

df.MonocyteMarker <- data.frame(Gene=NA,Mean=NA,SD=NA,SEM=NA,Datatype=NA)
for (m in Neutrophil.Markers) {

  cat(m,"Classical.Monocyte","All\n")
  #cat("subsetting finished\n")
  df.tmp<-data.frame(Gene=m, Mean=mean(Classical.Monocyte@data[m,]), SD=sd(Classical.Monocyte@data[m,]),SEM= sem(Classical.Monocyte@data[m,]),Datatype=paste0("Classical.Monocyte",".all"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)
  
  cat(m,"Classical.Monocyte","HC-HCY\n")
  tmp.Seurat<-Classical.Monocyte.HC_HCY
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Classical.Monocyte",".HC-HCY"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)

  cat(m,"Classical.Monocyte","HF\n")
  tmp.Seurat<-Classical.Monocyte.HF
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Classical.Monocyte",".HF"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)
  
  #Intermediate.Monocyte
  cat(m,"Intermediate.Monocyte","All\n")
  #cat("subsetting finished\n")
  df.tmp<-data.frame(Gene=m, Mean=mean(Intermediate.Monocyte@data[m,]), SD=sd(Intermediate.Monocyte@data[m,]),SEM= sem(Intermediate.Monocyte@data[m,]),Datatype=paste0("Intermediate.Monocyte",".all"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)
  
  cat(m,"Intermediate.Monocyte","HC-HCY\n")
  tmp.Seurat<-Intermediate.Monocyte.HC_HCY
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Intermediate.Monocyte",".HC-HCY"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)

  cat(m,"Intermediate.Monocyte","HF\n")
  tmp.Seurat<-Intermediate.Monocyte.HF
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Intermediate.Monocyte",".HF"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)

  
  cat(m,"Non.Classical.Monocyte","All\n")
  #cat("subsetting finished\n")
  df.tmp<-data.frame(Gene=m, Mean=mean(Non.Classical.Monocyte@data[m,]), SD=sd(Non.Classical.Monocyte@data[m,]),SEM= sem(Non.Classical.Monocyte@data[m,]),Datatype=paste0("Non.Classical.Monocyte",".all"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)
  
  cat(m,"Non.Classical.Monocyte","HC-HCY\n")
  tmp.Seurat<-Non.Classical.Monocyte.HC_HCY
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Non.Classical.Monocyte",".HC-HCY"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)

  cat(m,"Non.Classical.Monocyte","HF\n")
  tmp.Seurat <- Non.Classical.Monocyte.HF
  df.tmp<-data.frame(Gene=m, Mean=mean(tmp.Seurat@data[m,]), SD=sd(tmp.Seurat@data[m,]),SEM= sem(tmp.Seurat@data[m,]),Datatype=paste0("Non.Classical.Monocyte",".HF"))
  df.MonocyteMarker <- rbind(df.MonocyteMarker,df.tmp)
  
}

df.MonocyteMarker <- remove_missing(df.MonocyteMarker)

write.csv2(x=df.MonocyteMarker, file = "/media/ATLAS_NGS_storage/Wesley/CircRes2019/MonocyteMarkers.csv")
write.csv2(x=df.MonocyteMarker, file = "/media/ATLAS_NGS_storage/Wesley/CircRes2019/NeutrophileMarkers.csv")

library(ggplot2)

#colors.bars <- c("#4a6e00","#4a6e20","#4a6e40","#a72900","#a72920","#a72940","#007cb0","#007cc0","#007cd0")
colors.bars <- c("coral", "red", "red3","darkmagenta", "darkorchid", "purple","dodgerblue", "blue", "navy")
p<-ggplot(data = df.MonocyteMarker, aes(x=Datatype, y=Mean, fill=Datatype))
p + geom_bar(stat = "identity", color="black", position = position_dodge(), ) + 
  geom_errorbar(aes(ymin=Mean-SEM, ymax=Mean+SEM), width=.2, ) + facet_wrap(~Gene, scales = "free") +
  theme(axis.text.x = element_blank()) + scale_fill_manual(values = colors.bars)
```


###3.1.8.) Run Monocle for both cohorts CircRes Samples
```{r}
library(Matrix)
#############################################
########        Export Seurat       #########
#############################################
#Extract data, phenotype data, and feature data from the SeuratObject

data <- as.sparseMatrix(as.matrix(BothCohortsMonocytes@scale.data))
sparsm

pd <- new('AnnotatedDataFrame', data = BothCohortsMonocytes@meta.data)

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle cds
BothCohortsMonocytesMonocle <- newCellDataSet(data,
                         phenoData = pd,
                         featureData = fd,
                         lowerDetectionLimit = 0.5,
                         expressionFamily = negbinomial.size())

#############################################
########      import to Monocle     #########
#############################################
library(monocle)
BothCohortsMonocytesMonocle <- estimateSizeFactors(BothCohortsMonocytesMonocle)
BothCohortsMonocytesMonocle <- estimateDispersions(BothCohortsMonocytesMonocle)
BothCohortsMonocytesMonocle <- detectGenes(BothCohortsMonocytesMonocle, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(BothCohortsMonocytesMonocle), num_cells_expressed >= 10))


# Log-transform each value in the expression matrix.
L <- log(exprs(BothCohortsMonocytesMonocle[expressed_genes,]))

# Standardize each gene, so that they are all on the same scale,
# Then melt the data with plyr so we can plot it easily
melted_dens_df <- melt(Matrix::t(scale(Matrix::t(L))))

# Plot the distribution of the standardized gene expression values.
qplot(value, geom = "density", data = melted_dens_df) +  stat_function(fun = dnorm, size = 0.5, color = 'red') +  xlab("Standardized log(FPKM)") +  ylab("Density")


pData(H1Monocle)$Total_mRNAs <- Matrix::colSums(exprs(H1Monocle))
pData(H1Monocle)


###########################################################
####    add parameters and conditions to Monocle    #######
###########################################################
patient_id <- c()
condition <- c()
for (barcode in rownames(pData(H1Monocle))){
  patient<-substr(barcode,(nchar(barcode)+1)-1,nchar(barcode)) #get last charackter of barcode, which represents the pattients
  patient_id<-c(patient_id,patient)
  if (patient %in% c('1','2','3','4')) { condition<-c(condition,"control") }
  else if (patient %in% c('5','6','7','8')) {condition<-c(condition,"heartFailure")}
  else {print("unexpected patient_id")}
}
pData(H1Monocle)$condition <- condition
pData(H1Monocle)$patient_id <- patient_id


##############################################
####    Quality control for Monocle    #######
##############################################
H1Monocle <- H1Monocle[,pData(H1Monocle)$Total_mRNAs < 1e6]
upper_bound <- 10^(mean(log10(pData(H1Monocle)$Total_mRNAs)) + 2*sd(log10(pData(H1Monocle)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(H1Monocle)$Total_mRNAs)) -  2*sd(log10(pData(H1Monocle)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(H1Monocle), geom = "density") +  geom_vline(xintercept = lower_bound) + geom_vline(xintercept = upper_bound)

disp_table <- dispersionTable(H1Monocle)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 1.0)
H1Monocle <- setOrderingFilter(H1Monocle, unsup_clustering_genes$gene_id)
plot_ordering_genes(H1Monocle)


# HSMM@auxClusteringData[["tSNE"]]$variance_explained <- NULL
plot_pc_variance_explained(H1Monocle, return_all = F) # norm_method='log'

######################################################
####    unsupervised clustering for Monocle    #######
######################################################
H1Monocle <- reduceDimension(H1Monocle, max_components = 6, num_dim = 6, reduction_method = 'tSNE', residualModelFormulaStr = "~nUMI + num_genes_expressed", verbose = T)
H1Monocle <- clusterCells(H1Monocle, num_clusters = 6)
plot_cell_clusters(H1Monocle, 1, 2, color = "patient_id")
plot_cell_clusters(H1Monocle, 1, 2, color = "condition")
plot_cell_clusters(H1Monocle, 1, 2, color = "Cluster")


#detect differential genes between healthy and disease
diffGenesCondition <- differentialGeneTest(H1Monocle, fullModelFormulaStr = "~condition", cores = 15)
# Select genes that are significant at an FDR < 10%
sig_genes <- subset(diffGenesCondition, qval < 0.01)
sig_genes <- sig_genes[order(sig_genes$qval),]
head(sig_genes)

diffGenesControlDisease <- H1Monocle[row.names(subset(fData(H1Monocle), gene_short_name %in% c("IL1B","S100A9", "S100A8","CYP1B1", "CXCL8"))),]
plot_genes_jitter(diffGenesControlDisease, grouping = "condition", color_by = "condition", nrow= 1, ncol = 5, plot_trend = TRUE, relative_expr = T)

##########################################
####    create cell trajectorys    #######
##########################################
to_be_tested <- row.names(subset(fData(H1Monocle), gene_short_name %in% c("S100A9", "S100A8", "CYP1B1", "IL1B","CXCL8","CXCL3")))
cds_subset <- H1Monocle[to_be_tested,]

#select the top 1000 significant genes as the ordering genes.
H1_ordering_genes <-  row.names(diffGenesCondition)[order(diffGenesCondition$qval)][1:1000]
H1Monocle <-  setOrderingFilter(H1Monocle, ordering_genes = H1_ordering_genes)

H1Monocle <- reduceDimension(H1Monocle, method = 'DDRTree')

H1Monocle <- orderCells(H1Monocle)

GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$condition)[,"0"]
    return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}


H1Monocle <- orderCells(H1Monocle, root_state = GM_state(H1Monocle))

plot_cell_trajectory(H1Monocle, color_by = "condition")
plot_cell_trajectory(H1Monocle, color_by = "patient_id") + facet_wrap(~patient_id, nrow = 1)

#find trajectory specific genes
BEAM_res <- BEAM(H1Monocle, branch_point = 1, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res_all<-data.frame(BEAM_res,branchPoint=1)
BEAM_res <- BEAM(H1Monocle, branch_point = 2, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=2))
#BEAM_res <- BEAM(H1Monocle, branch_point = 3, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
#BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=3))
#BEAM_res <- BEAM(H1Monocle, branch_point = 4, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
#BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=4))

BEAM_res_all_filtered <- subset(BEAM_res_all,qval < 0.1)
setwd("/media/ATLAS_NGS_storage/Wesley/singlecell_Jan_2018/103305_aggrogated/outs/Monocle-Analysis")
write.csv2(BEAM_res_all_filtered, file="BranchSpecificGenes_04.11.18.csv")
BEAM_res <- BEAM(H1Monocle, cores = 9);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]
#plot Heatmap
plot_genes_branched_heatmap(H1Monocle[row.names(subset(BEAM_res,qval < 1e-7)),],branch_point = 1,num_clusters = 4, cores = 1,use_gene_short_name = T,show_rownames = T)


blast_genes <- row.names(subset(fData(H1Monocle), gene_short_name %in% c("S100A9", "S100A8", "CYP1B1", "IL1B","CXCL8","CXCL3")))
plot_genes_jitter(H1Monocle[blast_genes,], grouping = "condition", min_expr = 0.1)

cds_subset <- H1Monocle[blast_genes,]
plot_genes_in_pseudotime(cds_subset, color_by = "condition")


```





# _
## 3.2.) Second Run
###3.2.1.) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)
library(monocle)

require(scales)

logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.rmd.log"
if (file.exists(logFile)) {
  file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)
```


## 3.2.2.) Define static parameters
```{r}



#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/"
                     )
Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","HF-7"
                             )
```

## 3.2.3.) Impute the dataset
```{r}
#imputeData(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData, cluster = 20, ncores = 20)
```

## 3.2.4.) import and combine the raw cellranger counts with CCA
```{r}
CD31.circulatingCells <- combineSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)

CD31.circulatingCells <- FindVariableGenes(object = CD31.circulatingCells)
CD31.circulatingCells <- RunPCA(object = CD31.circulatingCells, features = VariableFeatures(object = CD31.circulatingCells), verbose = TRUE)
CD31.circulatingCells<-ProjectPCA(object = CD31.circulatingCells, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells <- FindClusters(object = CD31.circulatingCells, reduction.type = "pca", dims.use = 1:11, resolution = 0.6, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells)
CD31.circulatingCells <- RunTSNE(object = CD31.circulatingCells, dims.use = 1:11, do.label=TRUE, do.fast = TRUE)
```


## 3.2.5.)generate TSNE Plots
All Samples combined
```{r}
CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE, do.label = TRUE, label.size = 9)


CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, do.return=TRUE)
```

Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells@cell.names
CD31.circulatingCells<-AddMetaData(object = CD31.circulatingCells,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells<- SetAllIdent(object = CD31.circulatingCells, id = "condition")
TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.01, do.return=TRUE)
```

Separated by condition colored by cluster
```{r}
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
for (ident in unique(CD31.circulatingCells@meta.data$condition)){
  #cat(a,"\n")
  cellsToUse<-grep(ident, CD31.circulatingCells@cell.names, value = TRUE)
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, label.size = 4, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE)
  }
```

Separated by Patient colored by cluster
```{r}
library(stringr)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste0(outputFolder,"/TSNE/png/TSNE-",ident,".png")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 1,label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
  }
```

Colored by Mutations
```{r}
TSNEPlot(object = CD31.circulatingCells.s, label=T, pt.size = 2.1, group.by="orig.ident", colors.use = c("green","green","green","grey","blue","blue","blue","blue","blue","red","orange","orange","pink","orange","orange","orange","orange","pink","green","green"))
```


## 3.2.6.) Barplot of cell per cluster
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
library(stringr)
V<- CD31.circulatingCells@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:length(unique(CD31.circulatingCells@meta.data$res.0.6))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()


```

##save
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/" 
save(CD31.circulatingCells,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.RData"))
```

```{r}

load(file=paste0(outputFolder,"Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.RData"))
```

## 3.2.7.) DEG's
### 3.2.7.1.) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

generate vlnPlots
```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = din)

```


### 3.2.7.2.) DEGs HF vs CTRL-Aged
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.CtrlAgedvsHF.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "Ctrl-Aged", ident.2 = "HF")
# --> No DEGs found

CD31.circulatingCells.CtrlAgedvsHF.markers$geneName<-rownames(CD31.circulatingCells.CtrlAgedvsHF.markers)
CD31.circulatingCells.CtrlAgedvsHF.markers<-CD31.circulatingCells.CtrlAgedvsHF.markers[order(CD31.circulatingCells.CtrlAgedvsHF.markers$avg_logFC),]
topup<-head(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)
topdown<-tail(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)


write.csv(CD31.circulatingCells.CtrlAgedvsHF.markers,file = paste0(outputFolder,"ClusterSpecificMarkers.CtrlAged-vs-HF.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```


```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells_downregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "Ctrl-Aged", ident.2 = "HF", only.pos = TRUE,print.bar = T)
CD31.circulatingCells_upregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "HF", ident.2 = "Ctrl-Aged", only.pos = TRUE,print.bar = T)
head(CD31.circulatingCells_downregulated,n=9);head(CD31.circulatingCells_upregulated,n=9)
topDown<-head(CD31.circulatingCells_downregulated,n=6)
topUp<-head(CD31.circulatingCells_upregulated,n=6)

```

generate vlnPlots topUp colored by orig.ident
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")


VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[1],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[2],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[3],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[4],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[5],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[6],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[7],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[8],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[9],x.lab.rot = TRUE)

```


```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[1],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[2],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[3],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[4],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[5],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[6],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[7],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[8],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[9],x.lab.rot = TRUE)
```

```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topUp))
VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topDown))
```



### 3.2.7.3.) DEG Old vs Young 
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.YoungvsOld.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "Ctrl-Young", ident.2 = "Ctrl-Aged")
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(CD31.circulatingCells.markers,file = paste0(outputFolder,"ClusterSpecificMarkers.CtrlYoung-vs-CtrlAged.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

## 3.2.8.) Trajectory
```{r}
library(monocle)
```

### 3.2.8.1.) Import to Monocle
```{r}
require(monocle)
CD31.circulatingCells.Monocle <- importCDS(CD31.circulatingCells)
CD31.circulatingCells.Monocle <- estimateSizeFactors(CD31.circulatingCells.Monocle)
CD31.circulatingCells.Monocle <- estimateDispersions(CD31.circulatingCells.Monocle)
CD31.circulatingCells.Monocle <- detectGenes(CD31.circulatingCells.Monocle, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(CD31.circulatingCells.Monocle), num_cells_expressed >= 10))
```


```{r}
# Log-transform each value in the expression matrix.
L <- log(exprs(CD31.circulatingCells.Monocle[expressed_genes,]))

# Standardize each gene, so that they are all on the same scale,
# Then melt the data with plyr so we can plot it easily
melted_dens_df <- melt(Matrix::t(scale(Matrix::t(L))))

# Plot the distribution of the standardized gene expression values.
qplot(value, geom = "density", data = melted_dens_df) +  stat_function(fun = dnorm, size = 0.5, color = 'red') +  xlab("Standardized log(FPKM)") +  ylab("Density")
```

```{r}
pData(CD31.circulatingCells.Monocle)$Total_mRNAs <- Matrix::colSums(exprs(CD31.circulatingCells.Monocle))
pData(CD31.circulatingCells.Monocle)


###########################################################
####    add parameters and conditions to Monocle    #######
###########################################################
patient_id <- c()
condition <- c()
for (barcode in rownames(pData(CD31.circulatingCells.Monocle))){
  patient<-unlist(strsplit(barcode,split = "_"))[1]
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentCondition<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  patient_id<-c(patient_id,patient)
  condition<-c(condition,currentCondition)
}
pData(CD31.circulatingCells.Monocle)$condition <- condition
pData(CD31.circulatingCells.Monocle)$patient_id <- patient_id
```

Quality control for Monocle
```{r}
CD31.circulatingCells.Monocle <- CD31.circulatingCells.Monocle[,pData(CD31.circulatingCells.Monocle)$Total_mRNAs < 1e6]
upper_bound <- 10^(mean(log10(pData(CD31.circulatingCells.Monocle)$Total_mRNAs)) + 2*sd(log10(pData(CD31.circulatingCells.Monocle)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(CD31.circulatingCells.Monocle)$Total_mRNAs)) -  2*sd(log10(pData(CD31.circulatingCells.Monocle)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(CD31.circulatingCells.Monocle), geom = "density") +  geom_vline(xintercept = lower_bound) + geom_vline(xintercept = upper_bound)

disp_table <- dispersionTable(CD31.circulatingCells.Monocle)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 1.0)
CD31.circulatingCells.Monocle <- setOrderingFilter(CD31.circulatingCells.Monocle, unsup_clustering_genes$gene_id)
plot_ordering_genes(CD31.circulatingCells.Monocle)


# HSMM@auxClusteringData[["tSNE"]]$variance_explained <- NULL
plot_pc_variance_explained(CD31.circulatingCells.Monocle, return_all = F) # norm_method='log'
```

```{r}
######################################################
####    unsupervised clustering for Monocle    #######
######################################################
CD31.circulatingCells.Monocle <- reduceDimension(CD31.circulatingCells.Monocle, max_components = 3, num_dim = 6, reduction_method = 'tSNE', residualModelFormulaStr = "~nUMI + num_genes_expressed", verbose = T)
CD31.circulatingCells.Monocle <- clusterCells(CD31.circulatingCells.Monocle,num_clusters = 9)
plot_cell_clusters(CD31.circulatingCells.Monocle)

plot_cell_clusters(CD31.circulatingCells.Monocle, 1, 2, color = "patient_id")
plot_cell_clusters(CD31.circulatingCells.Monocle, 1, 2, color = "condition")
plot_cell_clusters(CD31.circulatingCells.Monocle, 1, 2, color = "Cluster")


#detect differential genes between healthy and disease
diffGenesCondition <- differentialGeneTest(CD31.circulatingCells.Monocle, fullModelFormulaStr = "~condition", cores = 15)
# Select genes that are significant at an FDR < 10%
sig_genes <- subset(diffGenesCondition, qval < 0.01)
sig_genes <- sig_genes[order(sig_genes$qval),]
head(sig_genes)

diffGenesControlDisease <- CD31.circulatingCells.Monocle[row.names(subset(fData(CD31.circulatingCells.Monocle), gene_short_name %in% c("IL1B","S100A9", "S100A8","CYP1B1", "CXCL8"))),]
plot_genes_jitter(diffGenesControlDisease, grouping = "condition", color_by = "condition", nrow= 1, ncol = 5, plot_trend = TRUE, relative_expr = T)
```

### 3.2.8.2.) Create Cell Trajectory
```{r}
#select the top 1000 significant genes as the ordering genes.
H1_ordering_genes <-  row.names(diffGenesCondition)[order(diffGenesCondition$qval)][1:1000]
CD31.circulatingCells.Monocle <-  setOrderingFilter(CD31.circulatingCells.Monocle, ordering_genes = H1_ordering_genes)
CD31.circulatingCells.Monocle <- reduceDimension(CD31.circulatingCells.Monocle, method = 'DDRTree')
CD31.circulatingCells.Monocle <- orderCells(CD31.circulatingCells.Monocle)
GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$condition)[,"Ctrl-Aged"]
    return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}
CD31.circulatingCells.Monocle <- orderCells(CD31.circulatingCells.Monocle, root_state = GM_state(CD31.circulatingCells.Monocle))
```

```{r}
plot_cell_trajectory(CD31.circulatingCells.Monocle, color_by = "condition")
plot_cell_trajectory(CD31.circulatingCells.Monocle, color_by = "patient_id") + facet_wrap(~patient_id, nrow = 1)
```

plot cake pie diagram
```{r}
trajectoryTable <- table(pData(CD31.circulatingCells.Monocle)$State, pData(CD31.circulatingCells.Monocle)$condition)
df<-as.data.frame(trajectoryTable)
ggplot(df, aes(x=factor(1), fill=Freq))+
  geom_bar(width = 1)+facet_grid(cols = vars(Var1)) +
  coord_polar("y")

```
```{r}
plot_cell_trajectory(CD31.circulatingCells.Monocle, color_by = "patient_id", markers = "CDH5") + facet_wrap(~patient_id, nrow = 1)

```




### 3.2.8.3.) Find trajectory specific genes
```{r}
BEAM_res <- BEAM(CD31.circulatingCells.Monocle, branch_point = 1, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res_all<-data.frame(BEAM_res,branchPoint=1)
BEAM_res <- BEAM(CD31.circulatingCells.Monocle, branch_point = 2, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=2))
#BEAM_res <- BEAM(H1Monocle, branch_point = 3, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
#BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=3))
#BEAM_res <- BEAM(H1Monocle, branch_point = 4, cores = 1);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
#BEAM_res_all<-rbind(BEAM_res_all,data.frame(BEAM_res,branchPoint=4))

BEAM_res_all_filtered <- subset(BEAM_res_all,qval < 0.1)
setwd("/media/ATLAS_NGS_storage/Wesley/singlecell_Jan_2018/103305_aggrogated/outs/Monocle-Analysis")
write.csv2(BEAM_res_all_filtered, file="BranchSpecificGenes_04.11.18.csv")
BEAM_res <- BEAM(H1Monocle, cores = 9);BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]
#plot Heatmap
plot_genes_branched_heatmap(CD31.circulatingCells.Monocle[row.names(subset(BEAM_res,qval < 1e-7)),],branch_point = 1,num_clusters = 4, cores = 1,use_gene_short_name = T,show_rownames = T)


blast_genes <- row.names(subset(fData(CD31.circulatingCells.Monocle), gene_short_name %in% c("S100A9", "S100A8", "CYP1B1", "IL1B","CXCL8","CXCL3")))
plot_genes_jitter(CD31.circulatingCells.Monocle[blast_genes,], grouping = "condition", min_expr = 0.1)

cds_subset <- CD31.circulatingCells.Monocle[blast_genes,]
plot_genes_in_pseudotime(cds_subset, color_by = "condition")


```


## 3.2.8.4.) Trajectory of CD14+ Cells (Classical Monocytes)
```{r}
FeaturePlot(CD31.circulatingCells, features.plot = c("CD14"))
CD14.Monocytes <- SubsetData(object = CD31.circulatingCells, ident.use = c("1","4","5","7"))

require(monocle)
CD14.Monocytes.Monocle <- importCDS(CD14.Monocytes)
CD14.Monocytes.Monocle <- estimateSizeFactors(CD14.Monocytes.Monocle)
CD14.Monocytes.Monocle <- estimateDispersions(CD14.Monocytes.Monocle)
CD14.Monocytes.Monocle <- detectGenes(CD14.Monocytes.Monocle, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(CD14.Monocytes.Monocle), num_cells_expressed >= 10))


pData(CD14.Monocytes.Monocle)$Total_mRNAs <- Matrix::colSums(exprs(CD14.Monocytes.Monocle))
pData(CD14.Monocytes.Monocle)


###########################################################
####    add parameters and conditions to Monocle    #######
###########################################################
patient_id <- c()
condition <- c()
for (barcode in rownames(pData(CD14.Monocytes.Monocle))){
  patient<-unlist(strsplit(barcode,split = "_"))[1]
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentCondition<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  patient_id<-c(patient_id,patient)
  condition<-c(condition,currentCondition)
}
pData(CD14.Monocytes.Monocle)$condition <- condition
pData(CD14.Monocytes.Monocle)$patient_id <- patient_id

CD14.Monocytes.Monocle <- CD14.Monocytes.Monocle[,pData(CD14.Monocytes.Monocle)$Total_mRNAs < 1e6]
upper_bound <- 10^(mean(log10(pData(CD14.Monocytes.Monocle)$Total_mRNAs)) + 2*sd(log10(pData(CD14.Monocytes.Monocle)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(CD14.Monocytes.Monocle)$Total_mRNAs)) -  2*sd(log10(pData(CD14.Monocytes.Monocle)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(CD14.Monocytes.Monocle), geom = "density") +  geom_vline(xintercept = lower_bound) + geom_vline(xintercept = upper_bound)

disp_table <- dispersionTable(CD14.Monocytes.Monocle)
unsup_clustering_genes <- subset(disp_table, mean_expression >= 1.0)
CD14.Monocytes.Monocle <- setOrderingFilter(CD14.Monocytes.Monocle, unsup_clustering_genes$gene_id)
plot_ordering_genes(CD14.Monocytes.Monocle)


# HSMM@auxClusteringData[["tSNE"]]$variance_explained <- NULL
plot_pc_variance_explained(CD14.Monocytes.Monocle, return_all = F) # norm_method='log'



######################################################
####    unsupervised clustering for Monocle    #######
######################################################
CD14.Monocytes.Monocle <- reduceDimension(CD14.Monocytes.Monocle, max_components = 3, num_dim = 6, reduction_method = 'tSNE', residualModelFormulaStr = "~nUMI + num_genes_expressed", verbose = T)
CD14.Monocytes.Monocle <- clusterCells(CD14.Monocytes.Monocle,num_clusters = 9)
plot_cell_clusters(CD14.Monocytes.Monocle)

plot_cell_clusters(CD14.Monocytes.Monocle, 1, 2, color = "patient_id")
plot_cell_clusters(CD14.Monocytes.Monocle, 1, 2, color = "condition")
plot_cell_clusters(CD14.Monocytes.Monocle, 1, 2, color = "Cluster")


#detect differential genes between healthy and disease
diffGenesCondition <- differentialGeneTest(CD14.Monocytes.Monocle, fullModelFormulaStr = "~condition", cores = 15)
# Select genes that are significant at an FDR < 10%
sig_genes <- subset(diffGenesCondition, qval < 0.01)
sig_genes <- sig_genes[order(sig_genes$qval),]
head(sig_genes)

diffGenesControlDisease <- CD14.Monocytes.Monocle[row.names(subset(fData(CD14.Monocytes.Monocle), gene_short_name %in% c("IL1B","S100A9", "S100A8","CYP1B1", "CXCL8"))),]
#plot_genes_jitter(diffGenesControlDisease)


#select the top 1000 significant genes as the ordering genes.
H1_ordering_genes <-  row.names(diffGenesCondition)[order(diffGenesCondition$qval)][1:1000]
CD14.Monocytes.Monocle <-  setOrderingFilter(CD14.Monocytes.Monocle, ordering_genes = H1_ordering_genes)
CD14.Monocytes.Monocle <- reduceDimension(CD14.Monocytes.Monocle, method = 'DDRTree')
CD14.Monocytes.Monocle <- orderCells(CD14.Monocytes.Monocle)

GM_state <- function(cds){
  if (length(unique(pData(cds)$State)) > 1){
    T0_counts <- table(pData(cds)$State, pData(cds)$condition)[,"Ctrl-Aged"]
    return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))]))
  } else {
    return (1)
  }
}
CD14.Monocytes.Monocle <- orderCells(CD14.Monocytes.Monocle, root_state = GM_state(CD14.Monocytes.Monocle))

plot_cell_trajectory(CD14.Monocytes.Monocle, color_by = "condition")
plot_cell_trajectory(CD14.Monocytes.Monocle, color_by = "patient_id") + facet_wrap(~patient_id, nrow = 1) + theme(legend.position = "none")
```


#_
###3.2.9.Circ Res Paper
```{r}
load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.RData")

#remove CTRL_YOUNG1, HF-1,HF-2,HF-7
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
CircRes_Cohort2<-SubsetData(CD31.circulatingCells,ident.use = c("Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3","HF-3","HF-4","HF-5","HF-6"))
```

```{r}
save(CircRes_Cohort2,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.circRes.RData")
load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.circRes.RData")
```

Recluster
```{r}
CircRes_Cohort2 <- FindVariableGenes(object = CircRes_Cohort2, do.recalc = T)
CircRes_Cohort2 <- RunPCA(object = CircRes_Cohort2, features = VariableFeatures(object = CircRes_Cohort2), verbose = TRUE)
CircRes_Cohort2<-ProjectPCA(object = CircRes_Cohort2, do.print = FALSE)
PCHeatmap(object = CircRes_Cohort2, pc.use = 1:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CircRes_Cohort2 <- FindClusters(object = CircRes_Cohort2, reduction.type = "pca", dims.use = 1:7, resolution = 0.6, save.SNN = TRUE, force.recalc = T)
PrintFindClustersParams(object = CircRes_Cohort2)
CircRes_Cohort2 <- RunTSNE(object = CircRes_Cohort2, dims.use = 1:7, do.label=TRUE)
```

Plots
```{r}
#CircRes_Cohort2<-SetAllIdent(object = CircRes_Cohort2, id = "res.0.6")
VlnPlot(CircRes_Cohort2, features.plot = "PECAM1")
FeaturePlot(object = CircRes_Cohort2,features.plot = c("FUT4","ELANE","CEACAM1"))
TSNEPlot(CircRes_Cohort2, do.label = T)

CircRes_Cohort2 <- SetAllIdent(object = CircRes_Cohort2, id = "condition")
VlnPlot(CircRes_Cohort2, features.plot = "TREM2")
FeaturePlot(object = CircRes_Cohort2,features.plot = c("TREM2"))
```

select CD31+ cells
```{r}
CircRes_Cohort2_CD31Pos <- SubsetData(CircRes_Cohort2, subset.name = "PECAM1", accept.low = 0.001)
CircRes_Cohort2_CD31Pos <- FindVariableGenes(object = CircRes_Cohort2_CD31Pos, do.recalc = T)
CircRes_Cohort2_CD31Pos <- RunPCA(object = CircRes_Cohort2_CD31Pos, features = VariableFeatures(object = CircRes_Cohort2_CD31Pos), verbose = TRUE)
CircRes_Cohort2_CD31Pos<-ProjectPCA(object = CircRes_Cohort2_CD31Pos, do.print = FALSE)
PCHeatmap(object = CircRes_Cohort2_CD31Pos, pc.use = 1:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CircRes_Cohort2_CD31Pos <- FindClusters(object = CircRes_Cohort2_CD31Pos, reduction.type = "pca", dims.use = 1:7, resolution = 0.6, save.SNN = TRUE, force.recalc = T)
PrintFindClustersParams(object = CircRes_Cohort2_CD31Pos)
CircRes_Cohort2_CD31Pos <- RunTSNE(object = CircRes_Cohort2_CD31Pos, dims.use = 1:4, do.label=TRUE)
```

Rename Samplenames
```{r}

newSamplename<-c()
for (sampleName in CircRes_Cohort2_CD31Pos@meta.data$orig.ident) {
  if (sampleName=="HF-1") {newSamplename<-c(newSamplename,"HF2.1")}
  if (sampleName=="HF-2") {newSamplename<-c(newSamplename,"HF2.2")}
  if (sampleName=="HF-3") {newSamplename<-c(newSamplename,"HF2.3")}
  if (sampleName=="HF-4") {newSamplename<-c(newSamplename,"HF2.4")}
  if (sampleName=="HF-5") {newSamplename<-c(newSamplename,"HF2.5")}
  if (sampleName=="HF-6") {newSamplename<-c(newSamplename,"HF2.6")}
  if (sampleName=="HF-7") {newSamplename<-c(newSamplename,"HF2.7")}
  if (sampleName=="HF-8") {newSamplename<-c(newSamplename,"HF1.1")}
  if (sampleName=="HF-9") {newSamplename<-c(newSamplename,"HF1.2")}
  if (sampleName=="HF-10") {newSamplename<-c(newSamplename,"HF1.3")}
  if (sampleName=="HF-11") {newSamplename<-c(newSamplename,"HF1.4")}
  if (sampleName=="Ctrl-Aged-1") {newSamplename<-c(newSamplename,"HC2.1")}
  if (sampleName=="Ctrl-Aged-2") {newSamplename<-c(newSamplename,"HC2.2")}
  if (sampleName=="Ctrl-Aged-3") {newSamplename<-c(newSamplename,"HC2.3")}
  if (sampleName=="Ctrl-Young-1") {newSamplename<-c(newSamplename,"HCY2.1")}
  if (sampleName=="Ctrl-Young-2") {newSamplename<-c(newSamplename,"HCY1.1")}
  if (sampleName=="Ctrl-Young-3") {newSamplename<-c(newSamplename,"HCY1.2")}
  if (sampleName=="Ctrl-Young-4") {newSamplename<-c(newSamplename,"HCY1.3")}
  if (sampleName=="Ctrl-Young-5") {newSamplename<-c(newSamplename,"HCY1.4")}
  if (sampleName=="Ctrl-CD14-Young-1") {newSamplename<-c(newSamplename,"HCY2.1_CD14")}
  else {cat("unknown Samplename: ", sampleName, "\n")}
}
table(newSamplename)
names(newSamplename)<-CircRes_Cohort2_CD31Pos@cell.names
CircRes_Cohort2_CD31Pos<-AddMetaData(object = CircRes_Cohort2_CD31Pos,metadata = data.frame(newSamplename),col.name = "newSamplenames")
table(CircRes_Cohort2_CD31Pos@meta.data$newSamplename)

rm(newSamplename)
```



Rename Samplenames for Circres Figures
```{r}

newSamplename.CircRes<-c()
for (sampleName in CircRes_Cohort2_CD31Pos@meta.data$orig.ident) {
  if (sampleName=="HF-3") {newSamplename.CircRes<-c(newSamplename.CircRes,"HF2.1")}
  if (sampleName=="HF-4") {newSamplename.CircRes<-c(newSamplename.CircRes,"HF2.2")}
  if (sampleName=="HF-5") {newSamplename.CircRes<-c(newSamplename.CircRes,"HF2.3")}
  if (sampleName=="HF-6") {newSamplename.CircRes<-c(newSamplename.CircRes,"HF2.4")}
  if (sampleName=="Ctrl-Aged-1") {newSamplename.CircRes<-c(newSamplename.CircRes,"HC2.1")}
  if (sampleName=="Ctrl-Aged-2") {newSamplename.CircRes<-c(newSamplename.CircRes,"HC2.2")}
  if (sampleName=="Ctrl-Aged-3") {newSamplename.CircRes<-c(newSamplename.CircRes,"HC2.3")}
  else {cat("unknown Samplename: ", sampleName, "\n")}
}
table(newSamplename.CircRes)
names(newSamplename.CircRes)<-CircRes_Cohort2_CD31Pos@cell.names
CircRes_Cohort2_CD31Pos<-AddMetaData(object = CircRes_Cohort2_CD31Pos,metadata = data.frame(newSamplename.CircRes),col.name = "newSamplenames.CircRes")
table(CircRes_Cohort2_CD31Pos@meta.data$newSamplename.CircRes)

rm(newSamplename.CircRes)
```



Plots
```{r}
#Neutrophil Markers
CircRes_Cohort2_CD31Pos <- SetAllIdent(object = CircRes_Cohort2_CD31Pos, id = "newSamplename.CircRes")

VlnPlot(CircRes_Cohort2_CD31Pos, features.plot = c("FUT4","ELANE"),x.lab.rot = 90)
FeaturePlot(object = CircRes_Cohort2_CD31Pos,features.plot = c("FUT4","ELANE"),cols.use = c("grey","red"))

VlnPlot(CircRes_Cohort2_CD31Pos, features.plot = c("PECAM1"),x.lab.rot = 90)


#Monocyte Markers
VlnPlot(CircRes_Cohort2_CD31Pos, features.plot = c("CD14","CD68","CD163"))
FeaturePlot(object = CircRes_Cohort2_CD31Pos,features.plot = c("CD14","CD68","CD163"))
TSNEPlot(CircRes_Cohort2_CD31Pos, do.label = T)
```

Get percentage of Neutrophil Marker positive cells
```{r}
df.NeutrophilPos <- data.frame()
CircRes_Cohort2_CD31Pos<-SetAllIdent(CircRes_Cohort2_CD31Pos,id = "orig.ident")
for (x in unique(CircRes_Cohort2_CD31Pos@meta.data$orig.ident)) {
  cat(x,"\n")
  tmp<-SubsetData(CircRes_Cohort2_CD31Pos,ident.use = x)
  df.NeutrophilPos[x,"TotalCells"]<-length(tmp@ident)
 
  n<-length(which(tmp@data["FUT4",]!=0))
  df.NeutrophilPos[x,"Number_Fut4_positive"]<-n
  df.NeutrophilPos[x,"Percentage_Fut4_positive"] <- (n/length(tmp@ident))*100
  
  n<-length(which(tmp@data["ELANE",]!=0))
  df.NeutrophilPos[x,"Number_ELANE_positive"]<-n
  df.NeutrophilPos[x,"Percentage_ELANE_positive"] <- (n/length(tmp@ident))*100

  n<-length(which(tmp@data["CEACAM1",]!=0))
  df.NeutrophilPos[x,"Number_CEACAM_positive"]<-n
  df.NeutrophilPos[x,"Percentage_CEACAM_positive"] <- (n/length(tmp@ident))*100
}
write.csv(x = df.NeutrophilPos, file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/PercentageNeutrophilMarkers.CircRes.csv")



length(which(CircRes_Cohort2_CD31Pos@data[rownames(CircRes_Cohort2_CD31Pos@data)[5718], ] != 0)) 
length(which(CircRes_Cohort2_CD31Pos@data["FUT4",]!=0))
median(CircRes_Cohort2_CD31Pos@data["FUT4",])


V<- CircRes_Cohort2_CD31Pos@meta.data

CircRes_Cohort2_CD31Pos.Fut4Pos <-SubsetData(CircRes_Cohort2_CD31Pos, subset.name = "FUT4", accept.low = 0.01)

V<- CircRes_Cohort2_CD31Pos.Fut4Pos@meta.data
orig.ident.ordered<-str_sort(unique(CircRes_Cohort2_CD31Pos.Fut4Pos@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:length(unique(CircRes_Cohort2_CD31Pos.Fut4Pos@meta.data$res.0.6))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()

```



DEG CTRL vs HF
```{r}
CircRes_Cohort2_CD31Pos <- SetAllIdent(object = CircRes_Cohort2_CD31Pos, id = "condition")
CircRes_Cohort2_CD31Pos.Markers.OldYoung <- FindAllMarkers(CircRes_Cohort2_CD31Pos)
```

```{r}

save(CircRes_Cohort2_CD31Pos,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.circRes.CD31+.RData")
load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Unimputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.circRes.CD31+.RData")
```



#_

#4.) Imputed Data (Separated By run)
##4.1.) First Run 
###4.1.1.) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)

require(scales)
library(Seurat)
logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/FirstRun//Wesley-PatientData-circulatingCells.Imputed.SeparatedByRun.FirstRun.rmd.log"
if (file.exists(logFile)) {
  #file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)
```


## 4.1.2.) Define static parameters
```{r}
#Static Parameters 
#CD34 Cells excluded for now
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-010",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-011",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-014",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-016",
#"/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-017",

#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy/103305-001-008/outs/filtered_feature_bc_matrix/")
Samplenames.PatientData <- c("Ctrl-Young-2","Ctrl-Young-3","Ctrl-Young-4",
                             "Ctrl-Young-5","HF-8","HF-9",
                             "HF-10","HF-11")
```

## 4.1.3. Import data normalise and Cluster
```{r}
CD31.circulatingCells.Cohort1.Imputed <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)



CD31.circulatingCells.Cohort1.Imputed <- FindVariableGenes(object = CD31.circulatingCells.Cohort1.Imputed)
CD31.circulatingCells.Cohort1.Imputed <- RunPCA(object = CD31.circulatingCells.Cohort1.Imputed, features = VariableFeatures(object = CD31.circulatingCells.Cohort1.Imputed), verbose = TRUE)
CD31.circulatingCells.Cohort1.Imputed<-ProjectPCA(object = CD31.circulatingCells.Cohort1.Imputed, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.Cohort1.Imputed, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.Cohort1.Imputed <- FindClusters(object = CD31.circulatingCells.Cohort1.Imputed, reduction.type = "pca", dims.use = 1:11, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.Cohort1.Imputed)
CD31.circulatingCells.Cohort1.Imputed <- RunTSNE(object = CD31.circulatingCells.Cohort1.Imputed, dims.use = 1:11, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.1, do.return=TRUE)
CD31.circulatingCells.Cohort1.Imputed<-SetAllIdent(CD31.circulatingCells.Cohort1.Imputed, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.1, do.label = T)
CD31.circulatingCells.Cohort1.Imputed<-SetAllIdent(CD31.circulatingCells.Cohort1.Imputed, id = "condition")
TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.1, do.label = T)

```




## 4.1.4. Generate TSNE's

Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells.Cohort1.Imputed@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells.Cohort1.Imputed@cell.names
CD31.circulatingCells.Cohort1.Imputed<-AddMetaData(object = CD31.circulatingCells.Cohort1.Imputed,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells.Cohort1.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort1.Imputed, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.01, do.return=TRUE)

CD31.circulatingCells.Cohort1.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort1.Imputed, id = "condition")
TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.01, do.return=TRUE)

# Load the "scales" package
require(scales)
my_color_palette <- hue_pal()(length(levels(CD31.circulatingCells.Cohort1.Imputed@ident)))




#generate TSNEPlots for the reviewer

TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.1, do.return=TRUE, cells.use = WhichCells(CD31.circulatingCells.Cohort1.Imputed,ident = "Ctrl-Young"))+ scale_color_manual(values = my_color_palette[1])
TSNEPlot(object = CD31.circulatingCells.Cohort1.Imputed, label=T, pt.size = 0.1, do.return=TRUE, cells.use = WhichCells(CD31.circulatingCells.Cohort1.Imputed,ident = "HF"))+ scale_color_manual(values = my_color_palette[2])
```

#### Suppl. Figure4 Cohort 1
All Cells
```{r}
#generate VLNPlot by hand for Stefanie 06.09.19
genes<-c("S100A8","S100A9","CXCL8","S100A12","VCAN","THBS1","CLEC7A","IL1B","CXCL2","EREG","CYP1B1","AC245014.3","PECAM1","CD14")

genes<-c("S100A8","S100A9","CXCL8","S100A12","VCAN","THBS1","CLEC7A","IL1B","CXCL2","EREG")


df.tmp<-data.frame(conditions=CD31.circulatingCells.Cohort1.Imputed@meta.data$condition)
for (i in genes) {
  df.tmp<-data.frame(df.tmp,CD31.circulatingCells.Cohort1.Imputed@data[i,])
}
colnames(df.tmp)<-c("conditions",genes)

library(reshape2)
df.melt<-melt(df.tmp,id.vars = "conditions")

ggplot(subset(df.melt, ),aes(x=conditions,y=value,fill=conditions)) + 
  geom_violin(scale = "width",) + facet_wrap(~variable, scales = "free", nrow = 2) + theme(strip.background = element_blank(), legend.position="none") + scale_fill_manual(values = c("red","blue")) + labs(title = "Cohort 1 Imputed All Cells")

```

```{r}
CD31.circulatingCells.Cohort1.Imputed.CD31Pos <- SubsetData(CD31.circulatingCells.Cohort1.Imputed, subset.name = "PECAM1", accept.low = 0.001)
CD31.circulatingCells.Cohort1.Imputed.CD31Pos <- FindVariableGenes(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos)
CD31.circulatingCells.Cohort1.Imputed.CD31Pos <- RunPCA(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos, features = VariableFeatures(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos), verbose = TRUE)
CD31.circulatingCells.Cohort1.Imputed.CD31Pos<-ProjectPCA(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos, pc.use = 1:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.Cohort1.Imputed.CD31Pos <- FindClusters(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos, reduction.type = "pca", dims.use = 1:7, resolution = 0.6, save.SNN = TRUE, force.recalc = T)
PrintFindClustersParams(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos)
CD31.circulatingCells.Cohort1.Imputed.CD31Pos <- RunTSNE(object = CD31.circulatingCells.Cohort1.Imputed.CD31Pos, dims.use = 1:10, do.label=TRUE)
TSNEPlot(CD31.circulatingCells.Cohort1.Imputed.CD31Pos)
```



CD31 positives
```{r}

#generate VLNPlot by hand for Stefanie 06.09.19
genes<-c("S100A8","S100A9","CXCL8","S100A12","VCAN","THBS1","CLEC7A","IL1B","CXCL2","EREG")

genes<-c("S100A8","S100A9","CXCL8","S100A12","THBS1","CLEC7A","VCAN","EREG")


df.tmp<-data.frame(conditions=CD31.circulatingCells.Cohort1.Imputed.CD31Pos@meta.data$condition)
for (i in genes) {
  df.tmp<-data.frame(df.tmp,CD31.circulatingCells.Cohort1.Imputed.CD31Pos@data[i,])
}
colnames(df.tmp)<-c("conditions",genes)

library(reshape2)
df.melt<-melt(df.tmp,id.vars = "conditions")

ggplot(subset(df.melt, ),aes(x=conditions,y=value,fill=conditions)) + 
  geom_violin(scale = "width",) + facet_wrap(~variable, scales = "free", nrow = 2) + theme(strip.background = element_blank(), legend.position="none") + scale_fill_manual(values = c("red","blue")) + labs(title = "Cohort 1 Imputed Monocytes")

```

All Cells
```{r}
#save
save(CD31.circulatingCells.Cohort1.Imputed, file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.Imputed.SeparatedByRun.FirstRun.RData")

load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.Imputed.SeparatedByRun.FirstRun.RData")
```

CD31 positive Cells
```{r}
#save
save(CD31.circulatingCells.Cohort1.Imputed.CD31Pos, file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.Imputed.SeparatedByRun.FirstRun.CD31Pos.RData")

load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/FirstRun/Wesley-PatientData-circulatingCells.Imputed.SeparatedByRun.FirstRun.RData")
```

=======
##4.2.) Second run

###4.2.1.) Import packages
```{r message=FALSE}
#load libraries
library(dplyr)
library(ggpubr)
library(monocle)

require(scales)

logFile="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.rmd.log"
if (file.exists(logFile)) {
  file.remove(logFile)
}
source("/home/david/scRNA-SEQ-ZMM/Import10X-HelperFunctions.R")
sink(file = logFile, append = TRUE, split = TRUE)
```


## 4.2.2.) Define static parameters
```{r}
outputFolder="/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/" 



#"HFpEF","Young_Control",Young_Control","ICM_HFrEF","Young_Control"
Path.PatientData.CD31.circulatingCells <- c(
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-001/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-002/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-003/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-004/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-005/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-006/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-007/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-008/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-009/outs/filtered_feature_bc_matrix/",
                     
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-012/outs/filtered_feature_bc_matrix/",
                     "/media/ATLAS_NGS_storage/Wesley/scRNA-SEQ_circulating-CD31+_HeartFailure_Healthy_Run2/103588-001-015/outs/filtered_feature_bc_matrix/"
                     )
Samplenames.PatientData <- c("HF-1","HF-2","HF-3",
                             "HF-4","HF-5","HF-6",
                             "Ctrl-Aged-1","Ctrl-Aged-2","Ctrl-Aged-3",
                             "Ctrl-Young-1","HF-7"
                             )
```


## 4.2.3.) Import an normalise Data
=======

```{r}
CD31.circulatingCells.Cohort2.Imputed <- combineScImputedSeuratObjectsCCA(pathways = Path.PatientData.CD31.circulatingCells, ids = Samplenames.PatientData)



#regress out batch effects 
mito.genes <- grep(pattern = "^MT-", x = rownames(x = CD31.circulatingCells.Cohort2.Imputed@data), value = TRUE)
percent.mito <- Matrix::colSums(CD31.circulatingCells.Cohort2.Imputed@raw.data[mito.genes, ])/Matrix::colSums(CD31.circulatingCells.Cohort2.Imputed@raw.data)
CD31.circulatingCells.Cohort2.Imputed <- AddMetaData(object = CD31.circulatingCells.Cohort2.Imputed, metadata = percent.mito, col.name = "percent.mito");rm(mito.genes, percent.mito)
CD31.circulatingCells.Cohort2.Imputed.filtered <- FilterCells(object = CD31.circulatingCells.Cohort2.Imputed, subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(Inf, 0.1))


GenePlot(object = CD31.circulatingCells.Cohort2.Imputed, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = CD31.circulatingCells.Cohort2.Imputed.filtered, gene1 = "nUMI", gene2 = "percent.mito")


CD31.circulatingCells.Cohort2.Imputed <- FindVariableGenes(object = CD31.circulatingCells.Cohort2.Imputed)
CD31.circulatingCells.Cohort2.Imputed <- RunPCA(object = CD31.circulatingCells.Cohort2.Imputed, features = VariableFeatures(object = CD31.circulatingCells.Cohort2.Imputed), verbose = TRUE)
CD31.circulatingCells.Cohort2.Imputed<-ProjectPCA(object = CD31.circulatingCells.Cohort2.Imputed, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.Cohort2.Imputed, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.Cohort2.Imputed <- FindClusters(object = CD31.circulatingCells.Cohort2.Imputed, reduction.type = "pca", dims.use = 1:11, resolution = 0.6, print.output = 0, save.SNN = TRUE)
PrintFindClustersParams(object = CD31.circulatingCells.Cohort2.Imputed)
CD31.circulatingCells.Cohort2.Imputed <- RunTSNE(object = CD31.circulatingCells.Cohort2.Imputed, dims.use = 1:11, do.fast = TRUE)


TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE)
CD31.circulatingCells.Cohort2.Imputed<-SetAllIdent(CD31.circulatingCells.Cohort2.Imputed, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.label = T)
```

## 4.2.4.)generate TSNE Plots
=======





## 3.1.5.)generate TSNE Plots
```{r}
CD31.circulatingCells.Cohort2.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort2.Imputed, id = "res.0.6")
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE, do.label = TRUE, label.size = 9)


CD31.circulatingCells.Cohort2.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort2.Imputed, id = "orig.ident")
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE)

CD31.circulatingCells.Cohort2.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort2.Imputed, id = "condition")
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE)
```

Separated by condition
```{r}
condition<-c()
for (barcode in colnames(CD31.circulatingCells.Cohort2.Imputed@data)) {
  tmp<-unlist(strsplit(barcode,split = "-"))
  currentBarcode<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,currentBarcode)
  
}
table(condition)
names(condition)<-CD31.circulatingCells.Cohort2.Imputed@cell.names
CD31.circulatingCells.Cohort2.Imputed<-AddMetaData(object = CD31.circulatingCells.Cohort2.Imputed,metadata = condition,col.name = "condition");rm(condition)

CD31.circulatingCells.Cohort2.Imputed<- SetAllIdent(object = CD31.circulatingCells.Cohort2.Imputed, id = "condition")
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.01, do.return=TRUE)



#generate TSNEPlots for the reviewer
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE, cells.use = WhichCells(CD31.circulatingCells.Cohort2.Imputed,ident = "Ctrl-Aged")) + scale_color_manual(values = my_color_palette[1])
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE, cells.use = WhichCells(CD31.circulatingCells.Cohort2.Imputed,ident = "Ctrl-Young"))+ scale_color_manual(values = my_color_palette[2])
TSNEPlot(object = CD31.circulatingCells.Cohort2.Imputed, label=T, pt.size = 0.1, do.return=TRUE, cells.use = WhichCells(CD31.circulatingCells.Cohort2.Imputed,ident = "HF"))+ scale_color_manual(values = my_color_palette[3])
```

#### Suppl. Figure4 Cohort 2
All Cells
```{r}
#generate VLNPlot by hand for Stefanie 06.09.19
genes<-c("S100A8","S100A9","CXCL8","S100A12","VCAN","THBS1","CLEC7A","IL1B","CXCL2","EREG")

df.tmp<-data.frame(conditions=CD31.circulatingCells.Cohort2.Imputed@meta.data$condition)
for (i in genes) {
  df.tmp<-data.frame(df.tmp,CD31.circulatingCells.Cohort2.Imputed@data[i,])
}
colnames(df.tmp)<-c("conditions",genes)

library(reshape2)
df.melt<-melt(df.tmp,id.vars = "conditions")

ggplot(subset(df.melt, subset = conditions!="Ctrl-Young"),aes(x=conditions,y=value,fill=conditions)) + 
  geom_violin(scale = "width",) + facet_wrap(~variable, scales = "free", nrow = 2) + theme(strip.background = element_blank(), legend.position="none") + scale_fill_manual(values = c("red","blue")) + labs(title = "Cohort 2 Imputed All Cells")

```

```{r}
CD31.circulatingCells.Cohort2.Imputed.CD31Pos <- SubsetData(CD31.circulatingCells.Cohort2.Imputed, subset.name = "PECAM1", accept.low = 0.001)
CD31.circulatingCells.Cohort2.Imputed.CD31Pos <- FindVariableGenes(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos)
CD31.circulatingCells.Cohort2.Imputed.CD31Pos <- RunPCA(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos, features = VariableFeatures(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos), verbose = TRUE)
CD31.circulatingCells.Cohort2.Imputed.CD31Pos<-ProjectPCA(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos, do.print = FALSE)
PCHeatmap(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos, pc.use = 1:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
CD31.circulatingCells.Cohort2.Imputed.CD31Pos <- FindClusters(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos, reduction.type = "pca", dims.use = 1:7, resolution = 0.6, save.SNN = TRUE, force.recalc = T)
PrintFindClustersParams(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos)
CD31.circulatingCells.Cohort2.Imputed.CD31Pos <- RunTSNE(object = CD31.circulatingCells.Cohort2.Imputed.CD31Pos, dims.use = 1:10, do.label=TRUE)
```



CD31 positives
```{r}

#generate VLNPlot by hand for Stefanie 06.09.19
genes<-c("S100A8","S100A9","CXCL8","S100A12","VCAN","THBS1","CLEC7A","IL1B","CXCL2","EREG")

genes<-c("S100A8","S100A9","CXCL8","S100A12","THBS1","CLEC7A","VCAN","EREG")


df.tmp<-data.frame(conditions=CD31.circulatingCells.Cohort2.Imputed.CD31Pos@meta.data$condition)
for (i in genes) {
  df.tmp<-data.frame(df.tmp,CD31.circulatingCells.Cohort2.Imputed.CD31Pos@data[i,])
}
colnames(df.tmp)<-c("conditions",genes)

library(reshape2)
df.melt<-melt(df.tmp,id.vars = "conditions")

ggplot(subset(df.melt, subset = conditions!="Ctrl-Young"),aes(x=conditions,y=value,fill=conditions)) + 
  geom_violin(scale = "width",) + facet_wrap(~variable, scales = "free", nrow = 2) + theme(strip.background = element_blank(), legend.position="none") + scale_fill_manual(values = c("red","blue")) + labs(title = "Cohort 2 Imputed Monocytes")

```


Separated by condition colored by cluster
```{r}
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
for (ident in unique(CD31.circulatingCells@meta.data$condition)){
  #cat(a,"\n")
  cellsToUse<-grep(ident, CD31.circulatingCells@cell.names, value = TRUE)
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 0.1, label.size = 4, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE)
  }
```

Separated by Patient colored by cluster
```{r}
library(stringr)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "orig.ident")
idents<- str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
CD31.circulatingCells<-SetAllIdent(CD31.circulatingCells,id = "res.0.6")
cellsToUse<-c()
for (ident in idents){
  x<-paste(ident, "_" ,sep = "")
  cat(x,ident,"\n")
  cellsToUse<-grep(x, CD31.circulatingCells@cell.names, value = TRUE)
  #head(cellsToUse)
  filename <- paste0(outputFolder,"/TSNE/png/TSNE-",ident,".png")
  #png(filename = filename,width = 800, height = 800)
  #TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 2, label.size = 12, cells.use = cellsToUse, do.label = TRUE, plot.title=ident,coord.fixed=TRUE,no.legend = TRUE, vector.friendly=TRUE,png.arguments=c(10,10,100))
  TSNEPlot(object = CD31.circulatingCells, label=T, pt.size = 1,label.size = 6, cells.use = cellsToUse, do.label = TRUE, plot.title=ident)
  #dev.off()
  }
```

## 4.2.5.) Barplot of cell per cluster
=======
```{r fig.height=10, fig.width=15}
# Counting celltypes in timepoints
library(tidyr)

library(dplyr)
library(ggplot2)
library(scales)
library(Seurat)
library(stringr)
V<- CD31.circulatingCells@meta.data
orig.ident.ordered<-str_sort(unique(CD31.circulatingCells@meta.data$orig.ident),numeric = TRUE)
V$orig.ident<-factor(V$orig.ident,levels = orig.ident.ordered)
V$res.0.6<-factor(V$res.0.6,levels = c(0:length(unique(CD31.circulatingCells@meta.data$res.0.6))))

Summary.Celltypes <- V %>% count(orig.ident,res.0.6,.drop = FALSE) %>% group_by(orig.ident) %>%
  mutate(freq = n /sum(n)) %>% complete(res.0.6,fill = list(n=0,freq=0))

Summary.Celltypes$res.0.6 <- factor(Summary.Celltypes$res.0.6)
condition<-c()
for (x in Summary.Celltypes$orig.ident) {
  tmp<-unlist(strsplit(x,split = "-"))
  cx<-paste0(tmp[1:length(tmp)-1],collapse = "-")
  
  condition<-c(condition,cx)
  
}
Summary.Celltypes$condition<-condition

svg(filename = paste0(outputFolder,"/Barplot-CellsperClusterPerSample.svg"),width = 15, height = 10)
ggplot(Summary.Celltypes, aes(x=orig.ident, y= freq, fill= condition))+
  geom_col(width = 0.9, color = "black")+
  facet_wrap(~res.0.6, nrow = 4, scales = "free")+
  scale_y_continuous(name = "Percent per timepoint", labels = scales::percent_format())+
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust= 1, size = 8))
dev.off()


```

#Save File
```{r}
save(CD31.circulatingCells,file = paste0(outputFolder,"Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.RData"))
```


## 3.1.7.) DEG's
### 3.1.7.1.) Cluster specific markers
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "res.0.6")
CD31.circulatingCells.markers <- FindAllMarkers(object = CD31.circulatingCells, only.pos = TRUE)
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(dink,file = paste0(outputFolder,"ClusterSpecificMarkers.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

generate vlnPlots
```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = c("MALAT1"))

```


### 3.1.7.2.) DEGs HF vs CTRL-Aged
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.CtrlAgedvsHF.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "Ctrl-Aged", ident.2 = "HF")
# --> No DEGs found

CD31.circulatingCells.CtrlAgedvsHF.markers$geneName<-rownames(CD31.circulatingCells.CtrlAgedvsHF.markers)
CD31.circulatingCells.CtrlAgedvsHF.markers<-CD31.circulatingCells.CtrlAgedvsHF.markers[order(CD31.circulatingCells.CtrlAgedvsHF.markers$avg_logFC),]
topup<-head(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)
topdown<-tail(x = CD31.circulatingCells.CtrlAgedvsHF.markers, n=9)


write.csv(CD31.circulatingCells.CtrlAgedvsHF.markers,file = paste0(outputFolder,"ClusterSpecificMarkers.CtrlAged-vs-HF.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```


```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells_downregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "Ctrl-Aged", ident.2 = "HF", only.pos = TRUE,print.bar = T)
CD31.circulatingCells_upregulated <- FindMarkers(CD31.circulatingCells, ident.1 = "HF", ident.2 = "Ctrl-Aged", only.pos = TRUE,print.bar = T)
head(CD31.circulatingCells_downregulated,n=9);head(CD31.circulatingCells_upregulated,n=9)
topDown<-head(CD31.circulatingCells_downregulated,n=6)
topUp<-head(CD31.circulatingCells_upregulated,n=6)

```

generate vlnPlots topUp colored by orig.ident
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "orig.ident")


VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[1],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[2],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[3],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[4],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[5],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[6],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[7],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[8],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topup$geneName[9],x.lab.rot = TRUE)

```


```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[1],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[2],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[3],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[4],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[5],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[6],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[7],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[8],x.lab.rot = TRUE)
VlnPlot(object = CD31.circulatingCells, features.plot = topdown$geneName[9],x.lab.rot = TRUE)
```

```{r}
VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topUp))
VlnPlot(object = CD31.circulatingCells, features.plot = rownames(topDown))
```



### 3.1.7.3.) DEG Old vs Young 
```{r}
CD31.circulatingCells <- SetAllIdent(object = CD31.circulatingCells, id = "condition")
CD31.circulatingCells.YoungvsOld.markers <- FindMarkers(object = CD31.circulatingCells, only.pos = FALSE, ident.1 = "Ctrl-Young", ident.2 = "Ctrl-Aged")
# --> No DEGs found

dink<-CD31.circulatingCells.markers %>% group_by(cluster) %>% top_n(20, avg_logFC)
write.csv(CD31.circulatingCells.markers,file = paste0(outputFolder,"ClusterSpecificMarkers.CtrlYoung-vs-CtrlAged.Wesley-PatientData-circulatingCells.unimputed.SeparatedByRun.SecondRun.csv"))
```

#Save
```{r}
save(CD31.circulatingCells.Cohort2.Imputed,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.RData")

load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.RData")
CD31.circulatingCells
CD31.circulatingCells.Cohort2.Imputed <- CD31.circulatingCells

```

```{r}
save(CD31.circulatingCells.Cohort2.Imputed.CD31Pos,file = "/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.CDE31Pos.RData")

load("/media/ATLAS_NGS_storage/Wesley/AnalysisDavid-2019/Imputed/SeparatedByRun/SecondRun/Wesley-PatientData-circulatingCells.imputed.SeparatedByRun.SecondRun.CDE31Pos.RData")
```

